{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Introduction to Bioconductor for Sequence Data\n",
    "\n",
    "Bioconductor enables the analysis and comprehension of high- throughput genomic data. We have a vast number of packages that allow rigorous statistical analysis of large data while keeping technological artifacts in mind. Bioconductor helps users place their analytic results into biological context, with rich opportunities for visualization. Reproducibility is an important goal in Bioconductor analyses. Different types of analysis can be carried out using Bioconductor, for example\n",
    "\n",
    "- Sequencing : RNASeq, ChIPSeq, variants, copy number..\n",
    "- Microarrays: expression, SNP, …\n",
    "- Domain specific analysis : Flow cytometry, Proteomics ..\n",
    "\n",
    "## Sequencing Resources\n",
    "Here is a illustrative description elaborating the different file types at various stages in a typical analysis, with the package names (in pink boxes) that one will use for each stage.\n",
    "![workflow](http://bioconductor.org/help/workflows/sequencing/sequencepkg.png)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Bioconductor version 3.1 (BiocInstaller 1.18.5), ?biocLite for help\n",
      "A newer version of Bioconductor is available for this version of R,\n",
      "  ?BiocUpgrade for help\n",
      "BioC_mirror: http://bioconductor.org\n",
      "Using Bioconductor version 3.1 (BiocInstaller 1.18.5), R version 3.2.2.\n",
      "Old packages: 'BH', 'boot', 'car', 'caret', 'digest', 'evaluate', 'formatR',\n",
      "  'ggplot2', 'glmnet', 'gtable', 'htmltools', 'htmlwidgets', 'jsonlite',\n",
      "  'knitr', 'lme4', 'maps', 'Matrix', 'mgcv', 'munsell', 'nlme', 'nnet',\n",
      "  'quantreg', 'R6', 'rbokeh', 'Rcpp', 'RcppEigen', 'rmarkdown', 'scales',\n",
      "  'shiny', 'tidyr', 'TTR', 'xtable'\n"
     ]
    }
   ],
   "source": [
    "source(\"http://bioconductor.org/biocLite.R\")\n",
    "biocLite()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "BioC_mirror: http://bioconductor.org\n",
      "Using Bioconductor version 3.1 (BiocInstaller 1.18.5), R version 3.2.2.\n",
      "Installing package(s) ‘GenomicRanges’, ‘Biostrings’, ‘AnnotationHub’,\n",
      "  ‘Rsamtools’, ‘BSgenome.Hsapiens.UCSC.hg19’\n",
      "also installing the dependencies ‘lambda.r’, ‘futile.options’, ‘futile.logger’, ‘snow’, ‘BiocParallel’, ‘XML’, ‘RCurl’, ‘GenomicAlignments’, ‘curl’, ‘openssl’, ‘rtracklayer’, ‘XVector’, ‘zlibbioc’, ‘interactiveDisplayBase’, ‘httr’, ‘BSgenome’\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "The downloaded source packages are in\n",
      "\t‘/tmp/RtmpOiAWCm/downloaded_packages’\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Old packages: 'BH', 'boot', 'car', 'caret', 'digest', 'evaluate', 'formatR',\n",
      "  'ggplot2', 'glmnet', 'gtable', 'htmltools', 'htmlwidgets', 'jsonlite',\n",
      "  'knitr', 'lme4', 'maps', 'Matrix', 'mgcv', 'munsell', 'nlme', 'nnet',\n",
      "  'quantreg', 'R6', 'rbokeh', 'Rcpp', 'RcppEigen', 'rmarkdown', 'scales',\n",
      "  'shiny', 'tidyr', 'TTR', 'xtable'\n"
     ]
    }
   ],
   "source": [
    "biolib_list <- c(\"GenomicRanges\",\"Biostrings\",\"AnnotationHub\",\"Rsamtools\",\"BSgenome.Hsapiens.UCSC.hg19\")\n",
    "\n",
    "biocLite(biolib_list)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Ranges Infrastructure\n",
    "\n",
    "The GenomicRanges package allows us to associate a range of chromosome coordinates with a sequence name (e.g., chromosome) and a strand. Such genomic ranges are very useful for describing both data (e.g., the coordinates of aligned reads, called ChIP peaks, SNPs, or copy number variants) and annotations (e.g., gene models, Roadmap Epigenomics regulatory elements, known clinically relevant variants from dbSNP). GRanges is an object representing a vector of genomic locations and associated annotations. Each element in the vector is comprised of a sequence name, a range, a strand, and optional metadata (e.g. score, GC content, etc.)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 10 ranges and 2 metadata columns:\n",
       "       seqnames    ranges strand |     score                 GC\n",
       "          <Rle> <IRanges>  <Rle> | <integer>          <numeric>\n",
       "   [1]     chr1  [ 1,  5]      - |       101  0.475001451326534\n",
       "   [2]     chr1  [ 2,  6]      - |       102  0.392243402078748\n",
       "   [3]     chr1  [ 3,  7]      - |       103  0.868793999776244\n",
       "   [4]     chr2  [ 4,  8]      - |       104  0.374351662350819\n",
       "   [5]     chr2  [ 5,  9]      - |       105  0.335704056546092\n",
       "   [6]     chr2  [ 6, 10]      - |       106  0.105049430392683\n",
       "   [7]     chr3  [ 7, 11]      - |       107  0.339706626022235\n",
       "   [8]     chr3  [ 8, 12]      - |       108  0.872086700517684\n",
       "   [9]     chr3  [ 9, 13]      - |       109  0.532811628654599\n",
       "  [10]     chr3  [10, 14]      - |       110 0.0882476158440113\n",
       "  -------\n",
       "  seqinfo: 3 sequences from an unspecified genome; no seqlengths"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "library(GenomicRanges)\n",
    "GRanges(seqnames=Rle(c('chr1', 'chr2', 'chr3'), c(3, 3, 4)),\n",
    "      IRanges(1:10, width=5), strand='-',\n",
    "      score=101:110, GC = runif(10))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<table width=\"100%\" summary=\"page for GRanges-class {GenomicRanges}\"><tr><td>GRanges-class {GenomicRanges}</td><td style=\"text-align: right;\">R Documentation</td></tr></table>\n",
       "\n",
       "<h2>GRanges objects</h2>\n",
       "\n",
       "<h3>Description</h3>\n",
       "\n",
       "<p>The GRanges class is a container for the genomic locations and\n",
       "their associated annotations.\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>Details</h3>\n",
       "\n",
       "<p>GRanges is a vector of genomic locations and associated\n",
       "annotations. Each element in the vector is comprised of a sequence name,\n",
       "an interval, a strand, and optional metadata columns (e.g. score, GC\n",
       "content, etc.). This information is stored in four components:\n",
       "</p>\n",
       "\n",
       "<dl>\n",
       "<dt><code>seqnames</code></dt><dd><p>a 'factor' Rle object\n",
       "containing the sequence names.</p>\n",
       "</dd>\n",
       "<dt><code>ranges</code></dt><dd><p>an IRanges object containing\n",
       "the ranges.</p>\n",
       "</dd>\n",
       "<dt><code>strand</code></dt><dd><p>a 'factor' Rle object containing\n",
       "the strand information.</p>\n",
       "</dd>\n",
       "<dt><code>mcols</code></dt><dd><p>a DataFrame object\n",
       "containing the metadata columns. Columns cannot be named\n",
       "<code>\"seqnames\"</code>, <code>\"ranges\"</code>, <code>\"strand\"</code>,\n",
       "<code>\"seqlevels\"</code>, <code>\"seqlengths\"</code>, <code>\"isCircular\"</code>,\n",
       "<code>\"start\"</code>, <code>\"end\"</code>, <code>\"width\"</code>, or <code>\"element\"</code>.</p>\n",
       "</dd>\n",
       "<dt><code>seqinfo</code></dt><dd><p>a Seqinfo object containing information\n",
       "about the set of genomic sequences present in the GRanges object.</p>\n",
       "</dd>\n",
       "</dl>\n",
       "\n",
       "\n",
       "\n",
       "<h3>Constructor</h3>\n",
       "\n",
       "\n",
       "<dl>\n",
       "<dt></dt><dd>\n",
       "<p><code>GRanges(seqnames=Rle(), ranges=NULL, strand=NULL,\n",
       "                    ...,\n",
       "                    seqlengths=NULL, seqinfo=NULL)</code>:\n",
       "Creates a GRanges object.\n",
       "</p>\n",
       "\n",
       "<dl>\n",
       "<dt><code>seqnames</code></dt><dd>\n",
       "<p>An Rle object, character vector, or factor\n",
       "containing the sequence names.\n",
       "</p>\n",
       "</dd>\n",
       "<dt><code>ranges</code></dt><dd>\n",
       "<p>An IRanges object containing the ranges.\n",
       "</p>\n",
       "</dd>\n",
       "<dt><code>strand</code></dt><dd>\n",
       "<p>Rle object, character vector, or factor\n",
       "containing the strand information.\n",
       "</p>\n",
       "</dd>\n",
       "<dt><code>...</code></dt><dd>\n",
       "<p>Optional metadata columns. These columns cannot be named\n",
       "<code>\"start\"</code>, <code>\"end\"</code>, <code>\"width\"</code>, or\n",
       "<code>\"element\"</code>.\n",
       "</p>\n",
       "</dd>\n",
       "<dt><code>seqlengths</code></dt><dd>\n",
       "<p>An integer vector named with <code>levels(seqnames)</code> and containing\n",
       "the lengths (or NA) for each level in <code>levels(seqnames)</code>.\n",
       "</p>\n",
       "</dd>\n",
       "<dt><code>seqinfo</code></dt><dd>\n",
       "<p>A Seqinfo object containing allowed\n",
       "sequence names, lengths (or NA), and circularity flag, for each\n",
       "level in <code>levels(seqnames)</code>.\n",
       "</p>\n",
       "</dd>\n",
       "</dl>\n",
       "\n",
       "<p>If <code>ranges</code> is not supplied and/or NULL then the constructor\n",
       "proceeds in 2 steps:\n",
       "</p>\n",
       "\n",
       "<ol>\n",
       "<li><p> An initial GRanges object is created with\n",
       "<code>as(seqnames, \"GRanges\")</code>.\n",
       "</p>\n",
       "</li>\n",
       "<li><p> Then this GRanges object is updated according to whatever\n",
       "non-NULL remaining arguments were passed to the call to\n",
       "<code>GRanges()</code>.\n",
       "</p>\n",
       "</li></ol>\n",
       "\n",
       "<p>Because of this behavior, <code>GRanges(x)</code> is equivalent to\n",
       "<code>as(x, \"GRanges\")</code>.\n",
       "</p>\n",
       "</dd>\n",
       "</dl>\n",
       "\n",
       "\n",
       "\n",
       "<h3>Coercion</h3>\n",
       "\n",
       "<p>In the code snippets below, <code>x</code> is a GRanges object.\n",
       "</p>\n",
       "\n",
       "<dl>\n",
       "<dt></dt><dd>\n",
       "<p><code>as(from, \"GRanges\")</code>: Creates a GRanges object from a character\n",
       "vector, a factor, or a RangedData, RangesList, RleList, or RleViewsList\n",
       "object.\n",
       "</p>\n",
       "<p>When <code>from</code> is a vector (or factor), each element in it must\n",
       "represent a genomic range in format <code>chr1:2501-2800</code>\n",
       "(unstranded range) or <code>chr1:2501-2800:+</code> (stranded range).\n",
       "<code>..</code> is also supported as a separator between the start and end\n",
       "positions. Strand can be <code>+</code>, <code>-</code>, <code>*</code>, or missing.\n",
       "The names on <code>from</code> are propagated to the returned GRanges object.\n",
       "See <code>as.character()</code> and <code>as.factor()</code> below for the\n",
       "reverse transformations.\n",
       "</p>\n",
       "<p>Coercing a data.frame or DataFrame into a GRanges object is also\n",
       "supported. See <code>makeGRangesFromDataFrame</code> for the details.\n",
       "</p>\n",
       "</dd>\n",
       "<dt></dt><dd>\n",
       "<p><code>as(from, \"RangedData\")</code>:\n",
       "Creates a RangedData object from a GRanges\n",
       "object. The <code>strand</code> and metadata columns become columns\n",
       "in the result. The <code>seqlengths(from)</code>, <code>isCircular(from)</code>,\n",
       "and <code>genome(from)</code> vectors are stored in the metadata columns\n",
       "of <code>ranges(rd)</code>.\n",
       "</p>\n",
       "</dd>\n",
       "<dt></dt><dd>\n",
       "<p><code>as(from, \"RangesList\")</code>:\n",
       "Creates a RangesList object from a GRanges\n",
       "object. The <code>strand</code> and metadata columns become <em>inner</em>\n",
       "metadata columns (i.e. metadata columns on the ranges).\n",
       "The <code>seqlengths(from)</code>, <code>isCircular(from)</code>, and\n",
       "<code>genome(from)</code> vectors become the metadata columns.\n",
       "</p>\n",
       "</dd>\n",
       "<dt></dt><dd>\n",
       "<p><code>as.character(x, ignore.strand=FALSE)</code>:\n",
       "Turn GRanges object <code>x</code> into a character vector where each\n",
       "range in <code>x</code> is represented by a string in format\n",
       "<code>chr1:2501-2800:+</code>. If <code>ignore.strand</code> is TRUE or if\n",
       "<em>all</em> the ranges in <code>x</code> are unstranded (i.e. their strand\n",
       "is set to <code>*</code>), then all the strings in the output are in\n",
       "format <code>chr1:2501-2800</code>.\n",
       "</p>\n",
       "<p>The names on <code>x</code> are propagated to the returned character vector.\n",
       "Its metadata (<code>metadata(x)</code>) and metadata columns (<code>mcols(x)</code>)\n",
       "are ignored.\n",
       "</p>\n",
       "<p>See <code>as(from, \"GRanges\")</code> above for the reverse transformation.\n",
       "</p>\n",
       "</dd>\n",
       "<dt></dt><dd>\n",
       "<p><code>as.factor(x)</code>:\n",
       "Equivalent to\n",
       "</p>\n",
       "<pre>  factor(as.character(x), levels=as.character(sort(unique(x))))\n",
       "</pre>\n",
       "<p>See <code>as(from, \"GRanges\")</code> above for the reverse transformation.\n",
       "</p>\n",
       "<p>Note that <code>table(x)</code> is supported on a GRanges object. It is\n",
       "equivalent to, but much faster than, <code>table(as.factor(x))</code>.\n",
       "</p>\n",
       "</dd>\n",
       "<dt></dt><dd>\n",
       "<p><code>as.data.frame(x, row.names = NULL, optional = FALSE, ...)</code>:\n",
       "Creates a data.frame with columns <code>seqnames</code> (factor),\n",
       "<code>start</code> (integer), <code>end</code> (integer), <code>width</code> (integer),\n",
       "<code>strand</code> (factor), as well as the additional metadata columns\n",
       "stored in <code>mcols(x)</code>. Pass an explicit\n",
       "<code>stringsAsFactors=TRUE/FALSE</code> argument via <code>...</code> to\n",
       "override the default conversions for the metadata columns in\n",
       "<code>mcols(x)</code>.\n",
       "</p>\n",
       "</dd>\n",
       "</dl>\n",
       "\n",
       "<p>In the code snippets below, <code>x</code> is a Seqinfo\n",
       "object.\n",
       "</p>\n",
       "\n",
       "<dl>\n",
       "<dt></dt><dd>\n",
       "<p><code>as(x, \"GRanges\")</code>, <code>as(x, \"GenomicRanges\")</code>,\n",
       "<code>as(x, \"RangesList\")</code>: Turns Seqinfo object\n",
       "<code>x</code> (with no <code>NA</code> lengths) into a GRanges or RangesList.\n",
       "</p>\n",
       "</dd>\n",
       "</dl>\n",
       "\n",
       "\n",
       "\n",
       "<h3>Accessors</h3>\n",
       "\n",
       "<p>In the following code snippets, <code>x</code> is a GRanges object.\n",
       "</p>\n",
       "\n",
       "<dl>\n",
       "<dt></dt><dd>\n",
       "<p><code>length(x)</code>:\n",
       "Get the number of elements.\n",
       "</p>\n",
       "</dd>\n",
       "<dt></dt><dd>\n",
       "<p><code>seqnames(x)</code>, <code>seqnames(x) &lt;- value</code>:\n",
       "Get or set the sequence names.\n",
       "<code>value</code> can be an Rle object, a character vector,\n",
       "or a factor.\n",
       "</p>\n",
       "</dd>\n",
       "<dt></dt><dd>\n",
       "<p><code>ranges(x)</code>, <code>ranges(x) &lt;- value</code>:\n",
       "Get or set the ranges. <code>value</code> can be a Ranges object.\n",
       "</p>\n",
       "</dd>\n",
       "<dt></dt><dd>\n",
       "<p><code>names(x)</code>, <code>names(x) &lt;- value</code>:\n",
       "Get or set the names of the elements.\n",
       "</p>\n",
       "</dd>\n",
       "<dt></dt><dd>\n",
       "<p><code>strand(x)</code>, <code>strand(x) &lt;- value</code>:\n",
       "Get or set the strand. <code>value</code> can be an Rle object, character\n",
       "vector, or factor.\n",
       "</p>\n",
       "</dd>\n",
       "<dt></dt><dd>\n",
       "<p><code>mcols(x, use.names=FALSE)</code>, <code>mcols(x) &lt;- value</code>:\n",
       "Get or set the metadata columns.\n",
       "If <code>use.names=TRUE</code> and the metadata columns are not <code>NULL</code>,\n",
       "then the names of <code>x</code> are propagated as the row names of the\n",
       "returned DataFrame object.\n",
       "When setting the metadata columns, the supplied value must be <code>NULL</code>\n",
       "or a data.frame-like object (i.e. DataTable or data.frame)\n",
       "object holding element-wise metadata.\n",
       "</p>\n",
       "</dd>\n",
       "<dt></dt><dd>\n",
       "<p><code>elementMetadata(x)</code>, <code>elementMetadata(x) &lt;- value</code>,\n",
       "<code>values(x)</code>, <code>values(x) &lt;- value</code>:\n",
       "Alternatives to <code>mcols</code> functions. Their use is discouraged.\n",
       "</p>\n",
       "</dd>\n",
       "<dt></dt><dd>\n",
       "<p><code>seqinfo(x)</code>, <code>seqinfo(x) &lt;- value</code>:\n",
       "Get or set the information about the underlying sequences.\n",
       "<code>value</code> must be a Seqinfo object.\n",
       "</p>\n",
       "</dd>\n",
       "<dt></dt><dd>\n",
       "<p><code>seqlevels(x)</code>, <code>seqlevels(x, force=FALSE) &lt;- value</code>:\n",
       "Get or set the sequence levels.\n",
       "<code>seqlevels(x)</code> is equivalent to <code>seqlevels(seqinfo(x))</code>\n",
       "or to <code>levels(seqnames(x))</code>, those 2 expressions being\n",
       "guaranteed to return identical character vectors on a GRanges object.\n",
       "<code>value</code> must be a character vector with no NAs.\n",
       "See <code>?seqlevels</code> for more information.\n",
       "</p>\n",
       "</dd>\n",
       "<dt></dt><dd>\n",
       "<p><code>seqlengths(x)</code>, <code>seqlengths(x) &lt;- value</code>:\n",
       "Get or set the sequence lengths.\n",
       "<code>seqlengths(x)</code> is equivalent to <code>seqlengths(seqinfo(x))</code>.\n",
       "<code>value</code> can be a named non-negative integer or numeric vector\n",
       "eventually with NAs.\n",
       "</p>\n",
       "</dd>\n",
       "<dt></dt><dd>\n",
       "<p><code>isCircular(x)</code>, <code>isCircular(x) &lt;- value</code>:\n",
       "Get or set the circularity flags.\n",
       "<code>isCircular(x)</code> is equivalent to <code>isCircular(seqinfo(x))</code>.\n",
       "<code>value</code> must be a named logical vector eventually with NAs.\n",
       "</p>\n",
       "</dd>\n",
       "<dt></dt><dd>\n",
       "<p><code>genome(x)</code>, <code>genome(x) &lt;- value</code>:\n",
       "Get or set the genome identifier or assembly name for each sequence.\n",
       "<code>genome(x)</code> is equivalent to <code>genome(seqinfo(x))</code>.\n",
       "<code>value</code> must be a named character vector eventually with NAs.\n",
       "</p>\n",
       "</dd>\n",
       "<dt></dt><dd>\n",
       "<p><code>seqlevelsStyle(x)</code>, <code>seqlevelsStyle(x) &lt;- value</code>:\n",
       "Get or set the seqname style for <code>x</code>.\n",
       "See the seqlevelsStyle generic getter and setter\n",
       "in the <span class=\"pkg\">GenomeInfoDb</span> package for more information.\n",
       "</p>\n",
       "</dd>\n",
       "<dt></dt><dd>\n",
       "<p><code>score(x), score(x) &lt;- value</code>: Get or set the &ldquo;score&rdquo;\n",
       "column from the element metadata.\n",
       "</p>\n",
       "</dd>\n",
       "<dt></dt><dd>\n",
       "<p><code>granges(x, use.mcols=FALSE)</code>: Gets a <code>GRanges</code> with\n",
       "only the range information from <code>x</code>, unless <code>use.mcols</code>\n",
       "is <code>TRUE</code>, in which case the metadata columns are also\n",
       "returned. Those columns will include any &quot;extra column slots&quot; if\n",
       "<code>x</code> is a specialized <code>GenomicRanges</code> derivative.\n",
       "</p>\n",
       "</dd>\n",
       "</dl>\n",
       "\n",
       "\n",
       "\n",
       "<h3>Ranges methods</h3>\n",
       "\n",
       "<p>In the following code snippets, <code>x</code> is a GRanges object.\n",
       "</p>\n",
       "\n",
       "<dl>\n",
       "<dt></dt><dd>\n",
       "<p><code>start(x)</code>, <code>start(x) &lt;- value</code>:\n",
       "Get or set <code>start(ranges(x))</code>.\n",
       "</p>\n",
       "</dd>\n",
       "<dt></dt><dd>\n",
       "<p><code>end(x)</code>, <code>end(x) &lt;- value</code>:\n",
       "Get or set <code>end(ranges(x))</code>.\n",
       "</p>\n",
       "</dd>\n",
       "<dt></dt><dd>\n",
       "<p><code>width(x)</code>, <code>width(x) &lt;- value</code>:\n",
       "Get or set <code>width(ranges(x))</code>.\n",
       "</p>\n",
       "</dd>\n",
       "</dl>\n",
       "\n",
       "\n",
       "\n",
       "<h3>Splitting and Combining</h3>\n",
       "\n",
       "<p>In the code snippets below, <code>x</code> is a GRanges object.\n",
       "</p>\n",
       "\n",
       "<dl>\n",
       "<dt></dt><dd>\n",
       "<p><code>append(x, values, after = length(x))</code>:\n",
       "Inserts the <code>values</code> into <code>x</code> at the position given by\n",
       "<code>after</code>, where <code>x</code> and <code>values</code> are of the same\n",
       "class.\n",
       "</p>\n",
       "</dd>\n",
       "<dt></dt><dd>\n",
       "<p><code>c(x, ...)</code>:\n",
       "Combines <code>x</code> and the GRanges objects in <code>...</code> together.\n",
       "Any object in <code>...</code> must belong to the same class as <code>x</code>,\n",
       "or to one of its subclasses, or must be <code>NULL</code>.\n",
       "The result is an object of the same class as <code>x</code>.\n",
       "</p>\n",
       "</dd>\n",
       "<dt></dt><dd>\n",
       "<p><code>c(x, ..., ignore.mcols=FALSE)</code>\n",
       "If the <code>GRanges</code> objects have metadata columns (represented as one\n",
       "DataFrame per object), each such DataFrame must have the\n",
       "same columns in order to combine successfully. In order to circumvent\n",
       "this restraint, you can pass in an <code>ignore.mcols=TRUE</code> argument\n",
       "which will combine all the objects into one and drop all of their\n",
       "metadata columns.\n",
       "</p>\n",
       "</dd>\n",
       "<dt></dt><dd>\n",
       "<p><code>split(x, f, drop=FALSE)</code>:\n",
       "Splits <code>x</code> according to <code>f</code> to create a\n",
       "GRangesList object.\n",
       "If <code>f</code> is a list-like object then <code>drop</code> is ignored\n",
       "and <code>f</code> is treated as if it was\n",
       "<code>rep(seq_len(length(f)), sapply(f, length))</code>,\n",
       "so the returned object has the same shape as <code>f</code> (it also\n",
       "receives the names of <code>f</code>).\n",
       "Otherwise, if <code>f</code> is not a list-like object, empty list\n",
       "elements are removed from the returned object if <code>drop</code> is\n",
       "<code>TRUE</code>.\n",
       "</p>\n",
       "</dd>\n",
       "</dl>\n",
       "\n",
       "\n",
       "\n",
       "<h3>Subsetting</h3>\n",
       "\n",
       "<p>In the code snippets below, <code>x</code> is a GRanges object.\n",
       "</p>\n",
       "\n",
       "<dl>\n",
       "<dt></dt><dd>\n",
       "<p><code>x[i, j]</code>, <code>x[i, j] &lt;- value</code>:\n",
       "Get or set elements <code>i</code> with optional metadata columns\n",
       "<code>mcols(x)[,j]</code>, where <code>i</code> can be missing; an NA-free\n",
       "logical, numeric, or character vector; or a 'logical' Rle object.\n",
       "</p>\n",
       "</dd>\n",
       "<dt></dt><dd>\n",
       "<p><code>x[i,j] &lt;- value</code>:\n",
       "Replaces elements <code>i</code> and optional metadata columns <code>j</code>\n",
       "with <code>value</code>.\n",
       "</p>\n",
       "</dd>\n",
       "<dt></dt><dd>\n",
       "<p><code>head(x, n = 6L)</code>:\n",
       "If <code>n</code> is non-negative, returns the first n elements of the\n",
       "GRanges object.\n",
       "If <code>n</code> is negative, returns all but the last <code>abs(n)</code> elements\n",
       "of the GRanges object.\n",
       "</p>\n",
       "</dd>\n",
       "<dt></dt><dd>\n",
       "<p><code>rep(x, times, length.out, each)</code>:\n",
       "Repeats the values in <code>x</code> through one of the following conventions:\n",
       "</p>\n",
       "\n",
       "<dl>\n",
       "<dt><code>times</code></dt><dd><p>Vector giving the number of times to repeat each\n",
       "element if of length <code>length(x)</code>, or to repeat the whole vector\n",
       "if of length 1.</p>\n",
       "</dd>\n",
       "<dt><code>length.out</code></dt><dd><p>Non-negative integer. The desired length of\n",
       "the output vector.</p>\n",
       "</dd>\n",
       "<dt><code>each</code></dt><dd><p>Non-negative integer.  Each element of <code>x</code> is\n",
       "repeated <code>each</code> times.</p>\n",
       "</dd>\n",
       "</dl>\n",
       "\n",
       "</dd>\n",
       "<dt></dt><dd>\n",
       "<p><code>subset(x, subset)</code>:\n",
       "Returns a new object of the same class as <code>x</code> made of the subset\n",
       "using logical vector <code>subset</code>, where missing values are taken as\n",
       "<code>FALSE</code>.\n",
       "</p>\n",
       "</dd>\n",
       "<dt></dt><dd>\n",
       "<p><code>tail(x, n = 6L)</code>:\n",
       "If <code>n</code> is non-negative, returns the last n elements of the\n",
       "GRanges object.\n",
       "If <code>n</code> is negative, returns all but the first <code>abs(n)</code> elements\n",
       "of the GRanges object.\n",
       "</p>\n",
       "</dd>\n",
       "<dt></dt><dd>\n",
       "<p><code>window(x, start = NA, end = NA, width = NA, frequency = NULL, delta = NULL, ...)</code>:\n",
       "Extracts the subsequence window from the GRanges object using:\n",
       "</p>\n",
       "\n",
       "<dl>\n",
       "<dt><code>start</code>, <code>end</code>, <code>width</code></dt><dd><p>The start, end, or width\n",
       "of the window. Two of the three are required.</p>\n",
       "</dd>\n",
       "<dt><code>frequency</code>, <code>delta</code></dt><dd><p>Optional arguments that specify\n",
       "the sampling frequency and increment within the window.</p>\n",
       "</dd>\n",
       "</dl>\n",
       "\n",
       "<p>In general, this is more efficient than using <code>\"[\"</code> operator.\n",
       "</p>\n",
       "</dd>\n",
       "<dt></dt><dd>\n",
       "<p><code>window(x, start = NA, end = NA, width = NA, keepLength = TRUE) &lt;- value</code>:\n",
       "Replaces the subsequence window specified on the left (i.e. the subsequence\n",
       "in <code>x</code> specified by <code>start</code>, <code>end</code> and <code>width</code>)\n",
       "by <code>value</code>.\n",
       "<code>value</code> must either be of class <code>class(x)</code>, belong to a subclass\n",
       "of <code>class(x)</code>, be coercible to <code>class(x)</code>, or be <code>NULL</code>.\n",
       "If <code>keepLength</code> is <code>TRUE</code>, the elements of <code>value</code> are\n",
       "repeated to create a GRanges object with the same number of elements\n",
       "as the width of the subsequence window it is replacing.\n",
       "If <code>keepLength</code> is <code>FALSE</code>, this replacement method can modify\n",
       "the length of <code>x</code>, depending on how the length of the left\n",
       "subsequence window compares to the length of <code>value</code>.\n",
       "</p>\n",
       "</dd>\n",
       "<dt></dt><dd>\n",
       "<p><code>x$name</code>, <code>x$name &lt;- value</code>:\n",
       "Shortcuts for <code>mcols(x)$name</code> and <code>mcols(x)$name &lt;- value</code>,\n",
       "respectively. Provided as a convenience, for GRanges objects *only*,\n",
       "and as the result of strong popular demand.\n",
       "Note that those methods are not consistent with the other <code>$</code>\n",
       "and <code>$&lt;-</code> methods in the IRanges/GenomicRanges infrastructure,\n",
       "and might confuse some users by making them believe that a GRanges\n",
       "object can be manipulated as a data.frame-like object.\n",
       "Therefore we recommend using them only interactively, and we discourage\n",
       "their use in scripts or packages. For the latter, use\n",
       "<code>mcols(x)$name</code> and <code>mcols(x)$name &lt;- value</code>, instead\n",
       "of <code>x$name</code> and <code>x$name &lt;- value</code>, respectively.\n",
       "</p>\n",
       "</dd>\n",
       "</dl>\n",
       "\n",
       "<p>Note that a GRanges object can be used to as a subscript to subset a\n",
       "list-like object that has names on it. In that case, the names on the\n",
       "list-like object are interpreted as sequence names.\n",
       "In the code snippets below, <code>x</code> is a list or List object with\n",
       "names on it, and the subscript <code>gr</code> is a GRanges object with all its\n",
       "seqnames being valid <code>x</code> names.\n",
       "</p>\n",
       "\n",
       "<dl>\n",
       "<dt></dt><dd>\n",
       "<p><code>x[gr]</code>:\n",
       "Return an object of the same class as <code>x</code> and <em>parallel</em>\n",
       "to <code>gr</code>. More precisely, it's conceptually doing:\n",
       "</p>\n",
       "<pre>  lapply(gr, function(gr1) x[[seqnames(gr1)]][ranges(gr1)])\n",
       "</pre>\n",
       "</dd>\n",
       "</dl>\n",
       "\n",
       "\n",
       "\n",
       "<h3>Other methods</h3>\n",
       "\n",
       "\n",
       "<dl>\n",
       "<dt></dt><dd>\n",
       "<p><code>show(x)</code>:\n",
       "By default the <code>show</code> method displays 5 head and 5 tail\n",
       "elements. This can be changed by setting the global options\n",
       "<code>showHeadLines</code> and <code>showTailLines</code>. If the object\n",
       "length is less than (or equal to) the sum of these 2 options\n",
       "plus 1, then the full object is displayed.\n",
       "Note that these options also affect the display of\n",
       "GAlignments and\n",
       "GAlignmentPairs objects (defined in\n",
       "the <span class=\"pkg\">GenomicAlignments</span> package), as well as other objects\n",
       "defined in the <span class=\"pkg\">IRanges</span> and <span class=\"pkg\">Biostrings</span> packages (e.g.\n",
       "IRanges and DNAStringSet objects).\n",
       "</p>\n",
       "</dd>\n",
       "</dl>\n",
       "\n",
       "\n",
       "\n",
       "<h3>Author(s)</h3>\n",
       "\n",
       "<p>P. Aboyoun and H. Pages</p>\n",
       "\n",
       "\n",
       "<h3>See Also</h3>\n",
       "\n",
       "\n",
       "<ul>\n",
       "<li> <p><code>makeGRangesFromDataFrame</code>\n",
       "</p>\n",
       "</li>\n",
       "<li> <p><code>seqinfo</code>\n",
       "</p>\n",
       "</li>\n",
       "<li> <p>GenomicRanges-comparison\n",
       "</p>\n",
       "</li>\n",
       "<li> <p>intra-range-methods\n",
       "</p>\n",
       "</li>\n",
       "<li> <p>inter-range-methods\n",
       "</p>\n",
       "</li>\n",
       "<li> <p>coverage-methods\n",
       "</p>\n",
       "</li>\n",
       "<li> <p>setops-methods\n",
       "</p>\n",
       "</li>\n",
       "<li> <p>findOverlaps-methods\n",
       "</p>\n",
       "</li>\n",
       "<li> <p>nearest-methods\n",
       "</p>\n",
       "</li>\n",
       "<li> <p><code>absoluteRanges</code>\n",
       "</p>\n",
       "</li>\n",
       "<li> <p><code>tileGenome</code>\n",
       "</p>\n",
       "</li>\n",
       "<li> <p>GRangesList-class\n",
       "</p>\n",
       "</li>\n",
       "<li> <p>Ranges-class\n",
       "</p>\n",
       "</li>\n",
       "<li> <p>Vector-class\n",
       "</p>\n",
       "</li>\n",
       "<li> <p>Rle-class\n",
       "</p>\n",
       "</li>\n",
       "<li> <p>DataFrame-class\n",
       "</p>\n",
       "</li></ul>\n",
       "\n",
       "\n",
       "\n",
       "<h3>Examples</h3>\n",
       "\n",
       "<pre>\n",
       "## ---------------------------------------------------------------------\n",
       "## CONSTRUCTION\n",
       "## ---------------------------------------------------------------------\n",
       "## Specifying the bare minimum i.e. seqnames and ranges only. The\n",
       "## GRanges object will have no names, no strand information, and no\n",
       "## metadata columns:\n",
       "gr0 &lt;- GRanges(Rle(c(\"chr2\", \"chr2\", \"chr1\", \"chr3\"), c(1, 3, 2, 4)),\n",
       "               IRanges(1:10, width=10:1))\n",
       "gr0\n",
       "\n",
       "## Specifying names, strand, metadata columns. They can be set on an\n",
       "## existing object:\n",
       "names(gr0) &lt;- head(letters, 10)\n",
       "strand(gr0) &lt;- Rle(strand(c(\"-\", \"+\", \"*\", \"+\", \"-\")), c(1, 2, 2, 3, 2))\n",
       "mcols(gr0)$score &lt;- 1:10\n",
       "mcols(gr0)$GC &lt;- seq(1, 0, length=10)\n",
       "gr0\n",
       "\n",
       "## ... or specified at construction time:\n",
       "gr &lt;- GRanges(Rle(c(\"chr2\", \"chr2\", \"chr1\", \"chr3\"), c(1, 3, 2, 4)),\n",
       "              IRanges(1:10, width=10:1, names=head(letters, 10)),\n",
       "              Rle(strand(c(\"-\", \"+\", \"*\", \"+\", \"-\")), c(1, 2, 2, 3, 2)),\n",
       "              score=1:10, GC=seq(1, 0, length=10))\n",
       "stopifnot(identical(gr0, gr))\n",
       "\n",
       "## Specifying the seqinfo. It can be set on an existing object:\n",
       "seqinfo &lt;- Seqinfo(paste0(\"chr\", 1:3), c(1000, 2000, 1500), NA, \"mock1\")\n",
       "seqinfo(gr0) &lt;- merge(seqinfo(gr0), seqinfo)\n",
       "seqlevels(gr0) &lt;- seqlevels(seqinfo)\n",
       "\n",
       "## ... or specified at construction time:\n",
       "gr &lt;- GRanges(Rle(c(\"chr2\", \"chr2\", \"chr1\", \"chr3\"), c(1, 3, 2, 4)),\n",
       "              IRanges(1:10, width=10:1, names=head(letters, 10)),\n",
       "              Rle(strand(c(\"-\", \"+\", \"*\", \"+\", \"-\")), c(1, 2, 2, 3, 2)),\n",
       "              score=1:10, GC=seq(1, 0, length=10),\n",
       "              seqinfo=seqinfo)\n",
       "stopifnot(identical(gr0, gr))\n",
       "\n",
       "## ---------------------------------------------------------------------\n",
       "## COERCION\n",
       "## ---------------------------------------------------------------------\n",
       "## From GRanges:\n",
       "as.character(gr)\n",
       "as.factor(gr)\n",
       "as.data.frame(gr)\n",
       "\n",
       "## From character to GRanges:\n",
       "x1 &lt;- \"chr2:56-125\"\n",
       "as(x1, \"GRanges\")\n",
       "as(rep(x1, 4), \"GRanges\")\n",
       "x2 &lt;- c(A=x1, B=\"chr1:25-30:-\")\n",
       "as(x2, \"GRanges\")\n",
       "\n",
       "## From data.frame to GRanges:\n",
       "df &lt;- data.frame(chrom=\"chr2\", start=11:15, end=20:24)\n",
       "gr3 &lt;- as(df, \"GRanges\")\n",
       "\n",
       "## Alternatively, coercion to GRanges can be done by just calling the\n",
       "## GRanges() constructor on the object to coerce:\n",
       "gr1 &lt;- GRanges(x1)  # same as as(x1, \"GRanges\")\n",
       "gr2 &lt;- GRanges(x2)  # same as as(x2, \"GRanges\")\n",
       "gr3 &lt;- GRanges(df)  # same as as(df, \"GRanges\")\n",
       "\n",
       "## Sanity checks:\n",
       "stopifnot(identical(as(x1, \"GRanges\"), gr1))\n",
       "stopifnot(identical(as(x2, \"GRanges\"), gr2))\n",
       "stopifnot(identical(as(df, \"GRanges\"), gr3))\n",
       "\n",
       "## ---------------------------------------------------------------------\n",
       "## SUMMARIZING ELEMENTS\n",
       "## ---------------------------------------------------------------------\n",
       "table(seqnames(gr))\n",
       "table(strand(gr))\n",
       "sum(width(gr))\n",
       "table(gr)\n",
       "summary(mcols(gr)[,\"score\"])\n",
       "\n",
       "## The number of lines displayed in the 'show' method are controlled\n",
       "## with two global options:\n",
       "longGR  &lt;- sample(gr, 25, replace=TRUE)\n",
       "longGR\n",
       "options(showHeadLines=7)\n",
       "options(showTailLines=2)\n",
       "longGR\n",
       "\n",
       "## Revert to default values\n",
       "options(showHeadLines=NULL)\n",
       "options(showTailLines=NULL)\n",
       "\n",
       "## ---------------------------------------------------------------------\n",
       "## RENAMING THE UNDERLYING SEQUENCES\n",
       "## ---------------------------------------------------------------------\n",
       "seqlevels(gr)\n",
       "seqlevels(gr) &lt;- sub(\"chr\", \"Chrom\", seqlevels(gr))\n",
       "gr\n",
       "seqlevels(gr) &lt;- sub(\"Chrom\", \"chr\", seqlevels(gr)) # revert\n",
       "\n",
       "## ---------------------------------------------------------------------\n",
       "## COMBINING OBJECTS\n",
       "## ---------------------------------------------------------------------\n",
       "gr2 &lt;- GRanges(seqnames=Rle(c('chr1', 'chr2', 'chr3'), c(3, 3, 4)),\n",
       "               IRanges(1:10, width=5),\n",
       "               strand='-',\n",
       "               score=101:110, GC=runif(10),\n",
       "               seqinfo=seqinfo)\n",
       "gr3 &lt;- GRanges(seqnames=Rle(c('chr1', 'chr2', 'chr3'), c(3, 4, 3)),\n",
       "               IRanges(101:110, width=10),\n",
       "               strand='-',\n",
       "               score=21:30,\n",
       "               seqinfo=seqinfo)\n",
       "some.gr &lt;- c(gr, gr2)\n",
       " \n",
       "## all.gr &lt;- c(gr, gr2, gr3) ## (This would fail)\n",
       "all.gr &lt;- c(gr, gr2, gr3, ignore.mcols=TRUE)\n",
       "\n",
       "## ---------------------------------------------------------------------\n",
       "## USING A GRANGES OBJECT AS A SUBSCRIPT TO SUBSET ANOTHER OBJECT\n",
       "## ---------------------------------------------------------------------\n",
       "## Subsetting *by* a GRanges subscript is supported only if the object\n",
       "## to subset is a named list-like object:\n",
       "x &lt;- RleList(chr1=101:120, chr2=2:-8, chr3=31:40)\n",
       "x[gr]\n",
       "</pre>\n",
       "\n",
       "<hr /><div style=\"text-align: center;\">[Package <em>GenomicRanges</em> version 1.20.8 ]</div>"
      ],
      "text/latex": [
       "\\inputencoding{utf8}\n",
       "\\HeaderA{GRanges-class}{GRanges objects}{GRanges.Rdash.class}\n",
       "\\aliasA{as.character,GenomicRanges-method}{GRanges-class}{as.character,GenomicRanges.Rdash.method}\n",
       "\\aliasA{as.data.frame,GenomicRanges-method}{GRanges-class}{as.data.frame,GenomicRanges.Rdash.method}\n",
       "\\aliasA{as.factor,GenomicRanges-method}{GRanges-class}{as.factor,GenomicRanges.Rdash.method}\n",
       "\\aliasA{c,GenomicRanges-method}{GRanges-class}{c,GenomicRanges.Rdash.method}\n",
       "\\aliasA{class:GenomicRanges}{GRanges-class}{class:GenomicRanges}\n",
       "\\aliasA{class:GRanges}{GRanges-class}{class:GRanges}\n",
       "\\aliasA{coerce,character,GenomicRanges-method}{GRanges-class}{coerce,character,GenomicRanges.Rdash.method}\n",
       "\\aliasA{coerce,character,GRanges-method}{GRanges-class}{coerce,character,GRanges.Rdash.method}\n",
       "\\aliasA{coerce,factor,GenomicRanges-method}{GRanges-class}{coerce,factor,GenomicRanges.Rdash.method}\n",
       "\\aliasA{coerce,factor,GRanges-method}{GRanges-class}{coerce,factor,GRanges.Rdash.method}\n",
       "\\aliasA{coerce,GenomicRanges,RangedData-method}{GRanges-class}{coerce,GenomicRanges,RangedData.Rdash.method}\n",
       "\\aliasA{coerce,GenomicRanges,RangesList-method}{GRanges-class}{coerce,GenomicRanges,RangesList.Rdash.method}\n",
       "\\aliasA{coerce,RangedData,GRanges-method}{GRanges-class}{coerce,RangedData,GRanges.Rdash.method}\n",
       "\\aliasA{coerce,RangesList,GRanges-method}{GRanges-class}{coerce,RangesList,GRanges.Rdash.method}\n",
       "\\aliasA{coerce,RleList,GRanges-method}{GRanges-class}{coerce,RleList,GRanges.Rdash.method}\n",
       "\\aliasA{coerce,RleViewsList,GRanges-method}{GRanges-class}{coerce,RleViewsList,GRanges.Rdash.method}\n",
       "\\aliasA{coerce,Seqinfo,GenomicRanges-method}{GRanges-class}{coerce,Seqinfo,GenomicRanges.Rdash.method}\n",
       "\\aliasA{coerce,Seqinfo,GRanges-method}{GRanges-class}{coerce,Seqinfo,GRanges.Rdash.method}\n",
       "\\aliasA{coerce,Seqinfo,RangesList-method}{GRanges-class}{coerce,Seqinfo,RangesList.Rdash.method}\n",
       "\\aliasA{elementMetadata<\\Rdash,GenomicRanges-method}{GRanges-class}{elementMetadata<.Rdash.,GenomicRanges.Rdash.method}\n",
       "\\aliasA{end<\\Rdash,GenomicRanges-method}{GRanges-class}{end<.Rdash.,GenomicRanges.Rdash.method}\n",
       "\\aliasA{end,GenomicRanges-method}{GRanges-class}{end,GenomicRanges.Rdash.method}\n",
       "\\aliasA{GenomicRanges}{GRanges-class}{GenomicRanges}\n",
       "\\aliasA{[<\\Rdash,GenomicRanges,ANY,ANY,ANY-method}{GRanges-class}{[<.Rdash.,GenomicRanges,ANY,ANY,ANY.Rdash.method}\n",
       "\\aliasA{[,GenomicRanges,ANY-method}{GRanges-class}{[,GenomicRanges,ANY.Rdash.method}\n",
       "\\aliasA{GenomicRanges-class}{GRanges-class}{GenomicRanges.Rdash.class}\n",
       "\\aliasA{\\$<\\Rdash,GenomicRanges-method}{GRanges-class}{.Rdol.<.Rdash.,GenomicRanges.Rdash.method}\n",
       "\\aliasA{\\$,GenomicRanges-method}{GRanges-class}{.Rdol.,GenomicRanges.Rdash.method}\n",
       "\\aliasA{GenomicRangesORGRangesList-class}{GRanges-class}{GenomicRangesORGRangesList.Rdash.class}\n",
       "\\aliasA{GenomicRangesORmissing-class}{GRanges-class}{GenomicRangesORmissing.Rdash.class}\n",
       "\\aliasA{GRanges}{GRanges-class}{GRanges}\n",
       "\\aliasA{GRanges}{GRanges-class}{GRanges}\n",
       "\\aliasA{granges,GenomicRanges-method}{GRanges-class}{granges,GenomicRanges.Rdash.method}\n",
       "\\aliasA{length,GenomicRanges-method}{GRanges-class}{length,GenomicRanges.Rdash.method}\n",
       "\\aliasA{[,list,GenomicRanges-method}{GRanges-class}{[,list,GenomicRanges.Rdash.method}\n",
       "\\aliasA{[,List,GenomicRanges-method}{GRanges-class}{[,List,GenomicRanges.Rdash.method}\n",
       "\\aliasA{names<\\Rdash,GenomicRanges-method}{GRanges-class}{names<.Rdash.,GenomicRanges.Rdash.method}\n",
       "\\aliasA{names,GenomicRanges-method}{GRanges-class}{names,GenomicRanges.Rdash.method}\n",
       "\\aliasA{ranges<\\Rdash,GenomicRanges-method}{GRanges-class}{ranges<.Rdash.,GenomicRanges.Rdash.method}\n",
       "\\aliasA{ranges,GRanges-method}{GRanges-class}{ranges,GRanges.Rdash.method}\n",
       "\\aliasA{score<\\Rdash,GenomicRanges-method}{GRanges-class}{score<.Rdash.,GenomicRanges.Rdash.method}\n",
       "\\aliasA{score,GenomicRanges-method}{GRanges-class}{score,GenomicRanges.Rdash.method}\n",
       "\\aliasA{seqinfo<\\Rdash,GenomicRanges-method}{GRanges-class}{seqinfo<.Rdash.,GenomicRanges.Rdash.method}\n",
       "\\aliasA{seqinfo,GRanges-method}{GRanges-class}{seqinfo,GRanges.Rdash.method}\n",
       "\\aliasA{seqnames<\\Rdash,GenomicRanges-method}{GRanges-class}{seqnames<.Rdash.,GenomicRanges.Rdash.method}\n",
       "\\aliasA{seqnames,GRanges-method}{GRanges-class}{seqnames,GRanges.Rdash.method}\n",
       "\\aliasA{show,GenomicRanges-method}{GRanges-class}{show,GenomicRanges.Rdash.method}\n",
       "\\aliasA{start<\\Rdash,GenomicRanges-method}{GRanges-class}{start<.Rdash.,GenomicRanges.Rdash.method}\n",
       "\\aliasA{start,GenomicRanges-method}{GRanges-class}{start,GenomicRanges.Rdash.method}\n",
       "\\aliasA{strand<\\Rdash,GenomicRanges,ANY-method}{GRanges-class}{strand<.Rdash.,GenomicRanges,ANY.Rdash.method}\n",
       "\\aliasA{strand,GRanges-method}{GRanges-class}{strand,GRanges.Rdash.method}\n",
       "\\aliasA{updateObject,GRanges-method}{GRanges-class}{updateObject,GRanges.Rdash.method}\n",
       "\\aliasA{width<\\Rdash,GenomicRanges-method}{GRanges-class}{width<.Rdash.,GenomicRanges.Rdash.method}\n",
       "\\aliasA{width,GenomicRanges-method}{GRanges-class}{width,GenomicRanges.Rdash.method}\n",
       "\\aliasA{window,GenomicRanges-method}{GRanges-class}{window,GenomicRanges.Rdash.method}\n",
       "%\n",
       "\\begin{Description}\\relax\n",
       "The GRanges class is a container for the genomic locations and\n",
       "their associated annotations.\n",
       "\\end{Description}\n",
       "%\n",
       "\\begin{Details}\\relax\n",
       "GRanges is a vector of genomic locations and associated\n",
       "annotations. Each element in the vector is comprised of a sequence name,\n",
       "an interval, a \\LinkA{strand}{strand}, and optional metadata columns (e.g. score, GC\n",
       "content, etc.). This information is stored in four components:\n",
       "\\begin{description}\n",
       "\n",
       "\\item[\\code{seqnames}] a 'factor' \\LinkA{Rle}{Rle} object\n",
       "containing the sequence names.\n",
       "\\item[\\code{ranges}] an \\LinkA{IRanges}{IRanges} object containing\n",
       "the ranges.\n",
       "\\item[\\code{strand}] a 'factor' \\LinkA{Rle}{Rle} object containing\n",
       "the \\LinkA{strand}{strand} information.\n",
       "\\item[\\code{mcols}] a \\LinkA{DataFrame}{DataFrame} object\n",
       "containing the metadata columns. Columns cannot be named\n",
       "\\code{\"seqnames\"}, \\code{\"ranges\"}, \\code{\"strand\"},\n",
       "\\code{\"seqlevels\"}, \\code{\"seqlengths\"}, \\code{\"isCircular\"},\n",
       "\\code{\"start\"}, \\code{\"end\"}, \\code{\"width\"}, or \\code{\"element\"}.\n",
       "\\item[\\code{seqinfo}] a \\LinkA{Seqinfo}{Seqinfo} object containing information\n",
       "about the set of genomic sequences present in the GRanges object.\n",
       "\n",
       "\\end{description}\n",
       "\n",
       "\\end{Details}\n",
       "%\n",
       "\\begin{Section}{Constructor}\n",
       "\\begin{description}\n",
       "\n",
       "\\item[] \n",
       "\\code{GRanges(seqnames=Rle(), ranges=NULL, strand=NULL,\n",
       "                    ...,\n",
       "                    seqlengths=NULL, seqinfo=NULL)}:\n",
       "Creates a GRanges object.\n",
       "\\begin{description}\n",
       "\n",
       "\\item[\\code{seqnames}] \n",
       "An \\LinkA{Rle}{Rle} object, character vector, or factor\n",
       "containing the sequence names.\n",
       "\n",
       "\\item[\\code{ranges}] \n",
       "An \\LinkA{IRanges}{IRanges} object containing the ranges.\n",
       "\n",
       "\\item[\\code{strand}] \n",
       "\\LinkA{Rle}{Rle} object, character vector, or factor\n",
       "containing the strand information.\n",
       "\n",
       "\\item[\\code{...}] \n",
       "Optional metadata columns. These columns cannot be named\n",
       "\\code{\"start\"}, \\code{\"end\"}, \\code{\"width\"}, or\n",
       "\\code{\"element\"}.\n",
       "\n",
       "\\item[\\code{seqlengths}] \n",
       "An integer vector named with \\code{levels(seqnames)} and containing\n",
       "the lengths (or NA) for each level in \\code{levels(seqnames)}.\n",
       "\n",
       "\\item[\\code{seqinfo}] \n",
       "A \\LinkA{Seqinfo}{Seqinfo} object containing allowed\n",
       "sequence names, lengths (or NA), and circularity flag, for each\n",
       "level in \\code{levels(seqnames)}.\n",
       "\n",
       "\n",
       "\\end{description}\n",
       "\n",
       "If \\code{ranges} is not supplied and/or NULL then the constructor\n",
       "proceeds in 2 steps:\n",
       "\\begin{enumerate}\n",
       "\n",
       "\\item An initial GRanges object is created with\n",
       "\\code{as(seqnames, \"GRanges\")}.\n",
       "\\item Then this GRanges object is updated according to whatever\n",
       "non-NULL remaining arguments were passed to the call to\n",
       "\\code{GRanges()}.\n",
       "\n",
       "\\end{enumerate}\n",
       "\n",
       "Because of this behavior, \\code{GRanges(x)} is equivalent to\n",
       "\\code{as(x, \"GRanges\")}.\n",
       "\n",
       "\n",
       "\\end{description}\n",
       "\n",
       "\\end{Section}\n",
       "%\n",
       "\\begin{Section}{Coercion}\n",
       "In the code snippets below, \\code{x} is a GRanges object.\n",
       "\n",
       "\\begin{description}\n",
       "\n",
       "\\item[] \n",
       "\\code{as(from, \"GRanges\")}: Creates a GRanges object from a character\n",
       "vector, a factor, or a RangedData, RangesList, RleList, or RleViewsList\n",
       "object.\n",
       "\n",
       "When \\code{from} is a vector (or factor), each element in it must\n",
       "represent a genomic range in format \\code{chr1:2501-2800}\n",
       "(unstranded range) or \\code{chr1:2501-2800:+} (stranded range).\n",
       "\\code{..} is also supported as a separator between the start and end\n",
       "positions. Strand can be \\code{+}, \\code{-}, \\code{*}, or missing.\n",
       "The names on \\code{from} are propagated to the returned GRanges object.\n",
       "See \\code{as.character()} and \\code{as.factor()} below for the\n",
       "reverse transformations.\n",
       "\n",
       "Coercing a data.frame or DataFrame into a GRanges object is also\n",
       "supported. See \\code{\\LinkA{makeGRangesFromDataFrame}{makeGRangesFromDataFrame}} for the details.\n",
       "\n",
       "\\item[] \n",
       "\\code{as(from, \"RangedData\")}:\n",
       "Creates a RangedData object from a GRanges\n",
       "object. The \\code{strand} and metadata columns become columns\n",
       "in the result. The \\code{seqlengths(from)}, \\code{isCircular(from)},\n",
       "and \\code{genome(from)} vectors are stored in the metadata columns\n",
       "of \\code{ranges(rd)}.\n",
       "\n",
       "\\item[] \n",
       "\\code{as(from, \"RangesList\")}:\n",
       "Creates a RangesList object from a GRanges\n",
       "object. The \\code{strand} and metadata columns become \\emph{inner}\n",
       "metadata columns (i.e. metadata columns on the ranges).\n",
       "The \\code{seqlengths(from)}, \\code{isCircular(from)}, and\n",
       "\\code{genome(from)} vectors become the metadata columns.\n",
       "\n",
       "\\item[] \n",
       "\\code{as.character(x, ignore.strand=FALSE)}:\n",
       "Turn GRanges object \\code{x} into a character vector where each\n",
       "range in \\code{x} is represented by a string in format\n",
       "\\code{chr1:2501-2800:+}. If \\code{ignore.strand} is TRUE or if\n",
       "\\emph{all} the ranges in \\code{x} are unstranded (i.e. their strand\n",
       "is set to \\code{*}), then all the strings in the output are in\n",
       "format \\code{chr1:2501-2800}.\n",
       "\n",
       "The names on \\code{x} are propagated to the returned character vector.\n",
       "Its metadata (\\code{metadata(x)}) and metadata columns (\\code{mcols(x)})\n",
       "are ignored.\n",
       "\n",
       "See \\code{as(from, \"GRanges\")} above for the reverse transformation.\n",
       "\n",
       "\\item[] \n",
       "\\code{as.factor(x)}:\n",
       "Equivalent to\n",
       "\\begin{alltt}  factor(as.character(x), levels=as.character(sort(unique(x))))\n",
       "\\end{alltt}\n",
       "\n",
       "See \\code{as(from, \"GRanges\")} above for the reverse transformation.\n",
       "\n",
       "Note that \\code{table(x)} is supported on a GRanges object. It is\n",
       "equivalent to, but much faster than, \\code{table(as.factor(x))}.\n",
       "\n",
       "\\item[] \n",
       "\\code{as.data.frame(x, row.names = NULL, optional = FALSE, ...)}:\n",
       "Creates a data.frame with columns \\code{seqnames} (factor),\n",
       "\\code{start} (integer), \\code{end} (integer), \\code{width} (integer),\n",
       "\\code{strand} (factor), as well as the additional metadata columns\n",
       "stored in \\code{mcols(x)}. Pass an explicit\n",
       "\\code{stringsAsFactors=TRUE/FALSE} argument via \\code{...} to\n",
       "override the default conversions for the metadata columns in\n",
       "\\code{mcols(x)}.\n",
       "\n",
       "\n",
       "\\end{description}\n",
       "\n",
       "\n",
       "In the code snippets below, \\code{x} is a \\LinkA{Seqinfo}{Seqinfo}\n",
       "object.\n",
       "\n",
       "\\begin{description}\n",
       "\n",
       "\\item[] \n",
       "\\code{as(x, \"GRanges\")}, \\code{as(x, \"GenomicRanges\")},\n",
       "\\code{as(x, \"RangesList\")}: Turns \\LinkA{Seqinfo}{Seqinfo} object\n",
       "\\code{x} (with no \\code{NA} lengths) into a GRanges or RangesList.\n",
       "\n",
       "\n",
       "\\end{description}\n",
       "\n",
       "\\end{Section}\n",
       "%\n",
       "\\begin{Section}{Accessors}\n",
       "In the following code snippets, \\code{x} is a GRanges object.\n",
       "\n",
       "\\begin{description}\n",
       "\n",
       "\\item[] \n",
       "\\code{length(x)}:\n",
       "Get the number of elements.\n",
       "\n",
       "\\item[] \n",
       "\\code{seqnames(x)}, \\code{seqnames(x) <- value}:\n",
       "Get or set the sequence names.\n",
       "\\code{value} can be an \\LinkA{Rle}{Rle} object, a character vector,\n",
       "or a factor.\n",
       "\n",
       "\\item[] \n",
       "\\code{ranges(x)}, \\code{ranges(x) <- value}:\n",
       "Get or set the ranges. \\code{value} can be a Ranges object.\n",
       "\n",
       "\\item[] \n",
       "\\code{names(x)}, \\code{names(x) <- value}:\n",
       "Get or set the names of the elements.\n",
       "\n",
       "\\item[] \n",
       "\\code{strand(x)}, \\code{strand(x) <- value}:\n",
       "Get or set the strand. \\code{value} can be an Rle object, character\n",
       "vector, or factor.\n",
       "\n",
       "\\item[] \n",
       "\\code{mcols(x, use.names=FALSE)}, \\code{mcols(x) <- value}:\n",
       "Get or set the metadata columns.\n",
       "If \\code{use.names=TRUE} and the metadata columns are not \\code{NULL},\n",
       "then the names of \\code{x} are propagated as the row names of the\n",
       "returned \\LinkA{DataFrame}{DataFrame} object.\n",
       "When setting the metadata columns, the supplied value must be \\code{NULL}\n",
       "or a data.frame-like object (i.e. \\LinkA{DataTable}{DataTable} or data.frame)\n",
       "object holding element-wise metadata.\n",
       "\n",
       "\\item[] \n",
       "\\code{elementMetadata(x)}, \\code{elementMetadata(x) <- value},\n",
       "\\code{values(x)}, \\code{values(x) <- value}:\n",
       "Alternatives to \\code{mcols} functions. Their use is discouraged.\n",
       "\n",
       "\\item[] \n",
       "\\code{seqinfo(x)}, \\code{seqinfo(x) <- value}:\n",
       "Get or set the information about the underlying sequences.\n",
       "\\code{value} must be a \\LinkA{Seqinfo}{Seqinfo} object.\n",
       "\n",
       "\\item[] \n",
       "\\code{seqlevels(x)}, \\code{seqlevels(x, force=FALSE) <- value}:\n",
       "Get or set the sequence levels.\n",
       "\\code{seqlevels(x)} is equivalent to \\code{seqlevels(seqinfo(x))}\n",
       "or to \\code{levels(seqnames(x))}, those 2 expressions being\n",
       "guaranteed to return identical character vectors on a GRanges object.\n",
       "\\code{value} must be a character vector with no NAs.\n",
       "See \\code{?\\LinkA{seqlevels}{seqlevels}} for more information.\n",
       "\n",
       "\\item[] \n",
       "\\code{seqlengths(x)}, \\code{seqlengths(x) <- value}:\n",
       "Get or set the sequence lengths.\n",
       "\\code{seqlengths(x)} is equivalent to \\code{seqlengths(seqinfo(x))}.\n",
       "\\code{value} can be a named non-negative integer or numeric vector\n",
       "eventually with NAs.\n",
       "\n",
       "\\item[] \n",
       "\\code{isCircular(x)}, \\code{isCircular(x) <- value}:\n",
       "Get or set the circularity flags.\n",
       "\\code{isCircular(x)} is equivalent to \\code{isCircular(seqinfo(x))}.\n",
       "\\code{value} must be a named logical vector eventually with NAs.\n",
       "\n",
       "\\item[] \n",
       "\\code{genome(x)}, \\code{genome(x) <- value}:\n",
       "Get or set the genome identifier or assembly name for each sequence.\n",
       "\\code{genome(x)} is equivalent to \\code{genome(seqinfo(x))}.\n",
       "\\code{value} must be a named character vector eventually with NAs.\n",
       "\n",
       "\\item[] \n",
       "\\code{seqlevelsStyle(x)}, \\code{seqlevelsStyle(x) <- value}:\n",
       "Get or set the seqname style for \\code{x}.\n",
       "See the \\LinkA{seqlevelsStyle}{seqlevelsStyle} generic getter and setter\n",
       "in the \\pkg{GenomeInfoDb} package for more information.\n",
       "\n",
       "\\item[] \n",
       "\\code{score(x), score(x) <- value}: Get or set the ``score''\n",
       "column from the element metadata.\n",
       "\n",
       "\\item[] \n",
       "\\code{granges(x, use.mcols=FALSE)}: Gets a \\code{GRanges} with\n",
       "only the range information from \\code{x}, unless \\code{use.mcols}\n",
       "is \\code{TRUE}, in which case the metadata columns are also\n",
       "returned. Those columns will include any \"extra column slots\" if\n",
       "\\code{x} is a specialized \\code{GenomicRanges} derivative.\n",
       "\n",
       "\n",
       "\\end{description}\n",
       "\n",
       "\\end{Section}\n",
       "%\n",
       "\\begin{Section}{Ranges methods}\n",
       "In the following code snippets, \\code{x} is a GRanges object.\n",
       "\n",
       "\\begin{description}\n",
       "\n",
       "\\item[] \n",
       "\\code{start(x)}, \\code{start(x) <- value}:\n",
       "Get or set \\code{start(ranges(x))}.\n",
       "\n",
       "\\item[] \n",
       "\\code{end(x)}, \\code{end(x) <- value}:\n",
       "Get or set \\code{end(ranges(x))}.\n",
       "\n",
       "\\item[] \n",
       "\\code{width(x)}, \\code{width(x) <- value}:\n",
       "Get or set \\code{width(ranges(x))}.\n",
       "\n",
       "\n",
       "\\end{description}\n",
       "\n",
       "\\end{Section}\n",
       "%\n",
       "\\begin{Section}{Splitting and Combining}\n",
       "In the code snippets below, \\code{x} is a GRanges object.\n",
       "\n",
       "\\begin{description}\n",
       "\n",
       "\\item[] \n",
       "\\code{append(x, values, after = length(x))}:\n",
       "Inserts the \\code{values} into \\code{x} at the position given by\n",
       "\\code{after}, where \\code{x} and \\code{values} are of the same\n",
       "class.\n",
       "\n",
       "\\item[] \n",
       "\\code{c(x, ...)}:\n",
       "Combines \\code{x} and the GRanges objects in \\code{...} together.\n",
       "Any object in \\code{...} must belong to the same class as \\code{x},\n",
       "or to one of its subclasses, or must be \\code{NULL}.\n",
       "The result is an object of the same class as \\code{x}.\n",
       "\n",
       "\\item[] \n",
       "\\code{c(x, ..., ignore.mcols=FALSE)}\n",
       "If the \\code{GRanges} objects have metadata columns (represented as one\n",
       "\\LinkA{DataFrame}{DataFrame} per object), each such \\LinkA{DataFrame}{DataFrame} must have the\n",
       "same columns in order to combine successfully. In order to circumvent\n",
       "this restraint, you can pass in an \\code{ignore.mcols=TRUE} argument\n",
       "which will combine all the objects into one and drop all of their\n",
       "metadata columns.\n",
       "\n",
       "\\item[] \n",
       "\\code{split(x, f, drop=FALSE)}:\n",
       "Splits \\code{x} according to \\code{f} to create a\n",
       "\\LinkA{GRangesList}{GRangesList} object.\n",
       "If \\code{f} is a list-like object then \\code{drop} is ignored\n",
       "and \\code{f} is treated as if it was\n",
       "\\code{rep(seq\\_len(length(f)), sapply(f, length))},\n",
       "so the returned object has the same shape as \\code{f} (it also\n",
       "receives the names of \\code{f}).\n",
       "Otherwise, if \\code{f} is not a list-like object, empty list\n",
       "elements are removed from the returned object if \\code{drop} is\n",
       "\\code{TRUE}.\n",
       "\n",
       "\n",
       "\\end{description}\n",
       "\n",
       "\\end{Section}\n",
       "%\n",
       "\\begin{Section}{Subsetting}\n",
       "In the code snippets below, \\code{x} is a GRanges object.\n",
       "\n",
       "\\begin{description}\n",
       "\n",
       "\\item[] \n",
       "\\code{x[i, j]}, \\code{x[i, j] <- value}:\n",
       "Get or set elements \\code{i} with optional metadata columns\n",
       "\\code{mcols(x)[,j]}, where \\code{i} can be missing; an NA-free\n",
       "logical, numeric, or character vector; or a 'logical' Rle object.\n",
       "\n",
       "\\item[] \n",
       "\\code{x[i,j] <- value}:\n",
       "Replaces elements \\code{i} and optional metadata columns \\code{j}\n",
       "with \\code{value}.\n",
       "\n",
       "\\item[] \n",
       "\\code{head(x, n = 6L)}:\n",
       "If \\code{n} is non-negative, returns the first n elements of the\n",
       "GRanges object.\n",
       "If \\code{n} is negative, returns all but the last \\code{abs(n)} elements\n",
       "of the GRanges object.\n",
       "\n",
       "\\item[] \n",
       "\\code{rep(x, times, length.out, each)}:\n",
       "Repeats the values in \\code{x} through one of the following conventions:\n",
       "\\begin{description}\n",
       "\n",
       "\\item[\\code{times}] Vector giving the number of times to repeat each\n",
       "element if of length \\code{length(x)}, or to repeat the whole vector\n",
       "if of length 1.\n",
       "\\item[\\code{length.out}] Non-negative integer. The desired length of\n",
       "the output vector.\n",
       "\\item[\\code{each}] Non-negative integer.  Each element of \\code{x} is\n",
       "repeated \\code{each} times.\n",
       "\n",
       "\\end{description}\n",
       "\n",
       "\n",
       "\\item[] \n",
       "\\code{subset(x, subset)}:\n",
       "Returns a new object of the same class as \\code{x} made of the subset\n",
       "using logical vector \\code{subset}, where missing values are taken as\n",
       "\\code{FALSE}.\n",
       "\n",
       "\\item[] \n",
       "\\code{tail(x, n = 6L)}:\n",
       "If \\code{n} is non-negative, returns the last n elements of the\n",
       "GRanges object.\n",
       "If \\code{n} is negative, returns all but the first \\code{abs(n)} elements\n",
       "of the GRanges object.\n",
       "\n",
       "\\item[] \n",
       "\\code{window(x, start = NA, end = NA, width = NA, frequency = NULL, delta = NULL, ...)}:\n",
       "Extracts the subsequence window from the GRanges object using:\n",
       "\\begin{description}\n",
       "\n",
       "\\item[\\code{start}, \\code{end}, \\code{width}] The start, end, or width\n",
       "of the window. Two of the three are required.\n",
       "\\item[\\code{frequency}, \\code{delta}] Optional arguments that specify\n",
       "the sampling frequency and increment within the window.\n",
       "\n",
       "\\end{description}\n",
       "\n",
       "In general, this is more efficient than using \\code{\"[\"} operator.\n",
       "\n",
       "\\item[] \n",
       "\\code{window(x, start = NA, end = NA, width = NA, keepLength = TRUE) <- value}:\n",
       "Replaces the subsequence window specified on the left (i.e. the subsequence\n",
       "in \\code{x} specified by \\code{start}, \\code{end} and \\code{width})\n",
       "by \\code{value}.\n",
       "\\code{value} must either be of class \\code{class(x)}, belong to a subclass\n",
       "of \\code{class(x)}, be coercible to \\code{class(x)}, or be \\code{NULL}.\n",
       "If \\code{keepLength} is \\code{TRUE}, the elements of \\code{value} are\n",
       "repeated to create a GRanges object with the same number of elements\n",
       "as the width of the subsequence window it is replacing.\n",
       "If \\code{keepLength} is \\code{FALSE}, this replacement method can modify\n",
       "the length of \\code{x}, depending on how the length of the left\n",
       "subsequence window compares to the length of \\code{value}.\n",
       "\n",
       "\\item[] \n",
       "\\code{x\\$name}, \\code{x\\$name <- value}:\n",
       "Shortcuts for \\code{mcols(x)\\$name} and \\code{mcols(x)\\$name <- value},\n",
       "respectively. Provided as a convenience, for GRanges objects *only*,\n",
       "and as the result of strong popular demand.\n",
       "Note that those methods are not consistent with the other \\code{\\$}\n",
       "and \\code{\\$<-} methods in the IRanges/GenomicRanges infrastructure,\n",
       "and might confuse some users by making them believe that a GRanges\n",
       "object can be manipulated as a data.frame-like object.\n",
       "Therefore we recommend using them only interactively, and we discourage\n",
       "their use in scripts or packages. For the latter, use\n",
       "\\code{mcols(x)\\$name} and \\code{mcols(x)\\$name <- value}, instead\n",
       "of \\code{x\\$name} and \\code{x\\$name <- value}, respectively.\n",
       "\n",
       "\n",
       "\\end{description}\n",
       "\n",
       "\n",
       "Note that a GRanges object can be used to as a subscript to subset a\n",
       "list-like object that has names on it. In that case, the names on the\n",
       "list-like object are interpreted as sequence names.\n",
       "In the code snippets below, \\code{x} is a list or \\LinkA{List}{List} object with\n",
       "names on it, and the subscript \\code{gr} is a GRanges object with all its\n",
       "seqnames being valid \\code{x} names.\n",
       "\n",
       "\\begin{description}\n",
       "\n",
       "\\item[] \n",
       "\\code{x[gr]}:\n",
       "Return an object of the same class as \\code{x} and \\emph{parallel}\n",
       "to \\code{gr}. More precisely, it's conceptually doing:\n",
       "\\begin{alltt}  lapply(gr, function(gr1) x[[seqnames(gr1)]][ranges(gr1)])\n",
       "\\end{alltt}\n",
       "\n",
       "\n",
       "\n",
       "\\end{description}\n",
       "\n",
       "\\end{Section}\n",
       "%\n",
       "\\begin{Section}{Other methods}\n",
       "\n",
       "\\begin{description}\n",
       "\n",
       "\\item[] \n",
       "\\code{show(x)}:\n",
       "By default the \\code{show} method displays 5 head and 5 tail\n",
       "elements. This can be changed by setting the global options\n",
       "\\code{showHeadLines} and \\code{showTailLines}. If the object\n",
       "length is less than (or equal to) the sum of these 2 options\n",
       "plus 1, then the full object is displayed.\n",
       "Note that these options also affect the display of\n",
       "\\LinkA{GAlignments}{GAlignments} and\n",
       "\\LinkA{GAlignmentPairs}{GAlignmentPairs} objects (defined in\n",
       "the \\pkg{GenomicAlignments} package), as well as other objects\n",
       "defined in the \\pkg{IRanges} and \\pkg{Biostrings} packages (e.g.\n",
       "\\LinkA{IRanges}{IRanges} and \\LinkA{DNAStringSet}{DNAStringSet} objects).\n",
       "\n",
       "\n",
       "\\end{description}\n",
       "\n",
       "\\end{Section}\n",
       "%\n",
       "\\begin{Author}\\relax\n",
       "P. Aboyoun and H. Pages\n",
       "\\end{Author}\n",
       "%\n",
       "\\begin{SeeAlso}\\relax\n",
       "\\begin{itemize}\n",
       "\n",
       "\\item \\code{\\LinkA{makeGRangesFromDataFrame}{makeGRangesFromDataFrame}}\n",
       "\\item \\code{\\LinkA{seqinfo}{seqinfo}}\n",
       "\\item \\LinkA{GenomicRanges-comparison}{GenomicRanges.Rdash.comparison}\n",
       "\\item \\LinkA{intra-range-methods}{intra.Rdash.range.Rdash.methods}\n",
       "\\item \\LinkA{inter-range-methods}{inter.Rdash.range.Rdash.methods}\n",
       "\\item \\LinkA{coverage-methods}{coverage.Rdash.methods}\n",
       "\\item \\LinkA{setops-methods}{setops.Rdash.methods}\n",
       "\\item \\LinkA{findOverlaps-methods}{findOverlaps.Rdash.methods}\n",
       "\\item \\LinkA{nearest-methods}{nearest.Rdash.methods}\n",
       "\\item \\code{\\LinkA{absoluteRanges}{absoluteRanges}}\n",
       "\\item \\code{\\LinkA{tileGenome}{tileGenome}}\n",
       "\\item \\LinkA{GRangesList-class}{GRangesList.Rdash.class}\n",
       "\\item \\LinkA{Ranges-class}{Ranges.Rdash.class}\n",
       "\\item \\LinkA{Vector-class}{Vector.Rdash.class}\n",
       "\\item \\LinkA{Rle-class}{Rle.Rdash.class}\n",
       "\\item \\LinkA{DataFrame-class}{DataFrame.Rdash.class}\n",
       "\n",
       "\\end{itemize}\n",
       "\n",
       "\\end{SeeAlso}\n",
       "%\n",
       "\\begin{Examples}\n",
       "\\begin{ExampleCode}\n",
       "## ---------------------------------------------------------------------\n",
       "## CONSTRUCTION\n",
       "## ---------------------------------------------------------------------\n",
       "## Specifying the bare minimum i.e. seqnames and ranges only. The\n",
       "## GRanges object will have no names, no strand information, and no\n",
       "## metadata columns:\n",
       "gr0 <- GRanges(Rle(c(\"chr2\", \"chr2\", \"chr1\", \"chr3\"), c(1, 3, 2, 4)),\n",
       "               IRanges(1:10, width=10:1))\n",
       "gr0\n",
       "\n",
       "## Specifying names, strand, metadata columns. They can be set on an\n",
       "## existing object:\n",
       "names(gr0) <- head(letters, 10)\n",
       "strand(gr0) <- Rle(strand(c(\"-\", \"+\", \"*\", \"+\", \"-\")), c(1, 2, 2, 3, 2))\n",
       "mcols(gr0)$score <- 1:10\n",
       "mcols(gr0)$GC <- seq(1, 0, length=10)\n",
       "gr0\n",
       "\n",
       "## ... or specified at construction time:\n",
       "gr <- GRanges(Rle(c(\"chr2\", \"chr2\", \"chr1\", \"chr3\"), c(1, 3, 2, 4)),\n",
       "              IRanges(1:10, width=10:1, names=head(letters, 10)),\n",
       "              Rle(strand(c(\"-\", \"+\", \"*\", \"+\", \"-\")), c(1, 2, 2, 3, 2)),\n",
       "              score=1:10, GC=seq(1, 0, length=10))\n",
       "stopifnot(identical(gr0, gr))\n",
       "\n",
       "## Specifying the seqinfo. It can be set on an existing object:\n",
       "seqinfo <- Seqinfo(paste0(\"chr\", 1:3), c(1000, 2000, 1500), NA, \"mock1\")\n",
       "seqinfo(gr0) <- merge(seqinfo(gr0), seqinfo)\n",
       "seqlevels(gr0) <- seqlevels(seqinfo)\n",
       "\n",
       "## ... or specified at construction time:\n",
       "gr <- GRanges(Rle(c(\"chr2\", \"chr2\", \"chr1\", \"chr3\"), c(1, 3, 2, 4)),\n",
       "              IRanges(1:10, width=10:1, names=head(letters, 10)),\n",
       "              Rle(strand(c(\"-\", \"+\", \"*\", \"+\", \"-\")), c(1, 2, 2, 3, 2)),\n",
       "              score=1:10, GC=seq(1, 0, length=10),\n",
       "              seqinfo=seqinfo)\n",
       "stopifnot(identical(gr0, gr))\n",
       "\n",
       "## ---------------------------------------------------------------------\n",
       "## COERCION\n",
       "## ---------------------------------------------------------------------\n",
       "## From GRanges:\n",
       "as.character(gr)\n",
       "as.factor(gr)\n",
       "as.data.frame(gr)\n",
       "\n",
       "## From character to GRanges:\n",
       "x1 <- \"chr2:56-125\"\n",
       "as(x1, \"GRanges\")\n",
       "as(rep(x1, 4), \"GRanges\")\n",
       "x2 <- c(A=x1, B=\"chr1:25-30:-\")\n",
       "as(x2, \"GRanges\")\n",
       "\n",
       "## From data.frame to GRanges:\n",
       "df <- data.frame(chrom=\"chr2\", start=11:15, end=20:24)\n",
       "gr3 <- as(df, \"GRanges\")\n",
       "\n",
       "## Alternatively, coercion to GRanges can be done by just calling the\n",
       "## GRanges() constructor on the object to coerce:\n",
       "gr1 <- GRanges(x1)  # same as as(x1, \"GRanges\")\n",
       "gr2 <- GRanges(x2)  # same as as(x2, \"GRanges\")\n",
       "gr3 <- GRanges(df)  # same as as(df, \"GRanges\")\n",
       "\n",
       "## Sanity checks:\n",
       "stopifnot(identical(as(x1, \"GRanges\"), gr1))\n",
       "stopifnot(identical(as(x2, \"GRanges\"), gr2))\n",
       "stopifnot(identical(as(df, \"GRanges\"), gr3))\n",
       "\n",
       "## ---------------------------------------------------------------------\n",
       "## SUMMARIZING ELEMENTS\n",
       "## ---------------------------------------------------------------------\n",
       "table(seqnames(gr))\n",
       "table(strand(gr))\n",
       "sum(width(gr))\n",
       "table(gr)\n",
       "summary(mcols(gr)[,\"score\"])\n",
       "\n",
       "## The number of lines displayed in the 'show' method are controlled\n",
       "## with two global options:\n",
       "longGR  <- sample(gr, 25, replace=TRUE)\n",
       "longGR\n",
       "options(showHeadLines=7)\n",
       "options(showTailLines=2)\n",
       "longGR\n",
       "\n",
       "## Revert to default values\n",
       "options(showHeadLines=NULL)\n",
       "options(showTailLines=NULL)\n",
       "\n",
       "## ---------------------------------------------------------------------\n",
       "## RENAMING THE UNDERLYING SEQUENCES\n",
       "## ---------------------------------------------------------------------\n",
       "seqlevels(gr)\n",
       "seqlevels(gr) <- sub(\"chr\", \"Chrom\", seqlevels(gr))\n",
       "gr\n",
       "seqlevels(gr) <- sub(\"Chrom\", \"chr\", seqlevels(gr)) # revert\n",
       "\n",
       "## ---------------------------------------------------------------------\n",
       "## COMBINING OBJECTS\n",
       "## ---------------------------------------------------------------------\n",
       "gr2 <- GRanges(seqnames=Rle(c('chr1', 'chr2', 'chr3'), c(3, 3, 4)),\n",
       "               IRanges(1:10, width=5),\n",
       "               strand='-',\n",
       "               score=101:110, GC=runif(10),\n",
       "               seqinfo=seqinfo)\n",
       "gr3 <- GRanges(seqnames=Rle(c('chr1', 'chr2', 'chr3'), c(3, 4, 3)),\n",
       "               IRanges(101:110, width=10),\n",
       "               strand='-',\n",
       "               score=21:30,\n",
       "               seqinfo=seqinfo)\n",
       "some.gr <- c(gr, gr2)\n",
       " \n",
       "## all.gr <- c(gr, gr2, gr3) ## (This would fail)\n",
       "all.gr <- c(gr, gr2, gr3, ignore.mcols=TRUE)\n",
       "\n",
       "## ---------------------------------------------------------------------\n",
       "## USING A GRANGES OBJECT AS A SUBSCRIPT TO SUBSET ANOTHER OBJECT\n",
       "## ---------------------------------------------------------------------\n",
       "## Subsetting *by* a GRanges subscript is supported only if the object\n",
       "## to subset is a named list-like object:\n",
       "x <- RleList(chr1=101:120, chr2=2:-8, chr3=31:40)\n",
       "x[gr]\n",
       "\\end{ExampleCode}\n",
       "\\end{Examples}"
      ],
      "text/plain": [
       "GRanges-class          package:GenomicRanges           R Documentation\n",
       "\n",
       "_\bG_\bR_\ba_\bn_\bg_\be_\bs _\bo_\bb_\bj_\be_\bc_\bt_\bs\n",
       "\n",
       "_\bD_\be_\bs_\bc_\br_\bi_\bp_\bt_\bi_\bo_\bn:\n",
       "\n",
       "     The GRanges class is a container for the genomic locations and\n",
       "     their associated annotations.\n",
       "\n",
       "_\bD_\be_\bt_\ba_\bi_\bl_\bs:\n",
       "\n",
       "     GRanges is a vector of genomic locations and associated\n",
       "     annotations. Each element in the vector is comprised of a sequence\n",
       "     name, an interval, a strand, and optional metadata columns (e.g.\n",
       "     score, GC content, etc.). This information is stored in four\n",
       "     components:\n",
       "\n",
       "     ‘seqnames’ a 'factor' Rle object containing the sequence names.\n",
       "\n",
       "     ‘ranges’ an IRanges object containing the ranges.\n",
       "\n",
       "     ‘strand’ a 'factor' Rle object containing the strand information.\n",
       "\n",
       "     ‘mcols’ a DataFrame object containing the metadata columns.\n",
       "          Columns cannot be named ‘\"seqnames\"’, ‘\"ranges\"’, ‘\"strand\"’,\n",
       "          ‘\"seqlevels\"’, ‘\"seqlengths\"’, ‘\"isCircular\"’, ‘\"start\"’,\n",
       "          ‘\"end\"’, ‘\"width\"’, or ‘\"element\"’.\n",
       "\n",
       "     ‘seqinfo’ a Seqinfo object containing information about the set of\n",
       "          genomic sequences present in the GRanges object.\n",
       "\n",
       "_\bC_\bo_\bn_\bs_\bt_\br_\bu_\bc_\bt_\bo_\br:\n",
       "\n",
       "‘GRanges(seqnames=Rle(), ranges=NULL, strand=NULL, ...,\n",
       "          seqlengths=NULL, seqinfo=NULL)’: Creates a GRanges object.\n",
       "\n",
       "          ‘seqnames’ An Rle object, character vector, or factor\n",
       "              containing the sequence names.\n",
       "\n",
       "          ‘ranges’ An IRanges object containing the ranges.\n",
       "\n",
       "          ‘strand’ Rle object, character vector, or factor containing\n",
       "              the strand information.\n",
       "\n",
       "          ‘...’ Optional metadata columns. These columns cannot be\n",
       "              named ‘\"start\"’, ‘\"end\"’, ‘\"width\"’, or ‘\"element\"’.\n",
       "\n",
       "          ‘seqlengths’ An integer vector named with ‘levels(seqnames)’\n",
       "              and containing the lengths (or NA) for each level in\n",
       "              ‘levels(seqnames)’.\n",
       "\n",
       "          ‘seqinfo’ A Seqinfo object containing allowed sequence names,\n",
       "              lengths (or NA), and circularity flag, for each level in\n",
       "              ‘levels(seqnames)’.\n",
       "\n",
       "          If ‘ranges’ is not supplied and/or NULL then the constructor\n",
       "          proceeds in 2 steps:\n",
       "\n",
       "           1. An initial GRanges object is created with ‘as(seqnames,\n",
       "              \"GRanges\")’.\n",
       "\n",
       "           2. Then this GRanges object is updated according to whatever\n",
       "              non-NULL remaining arguments were passed to the call to\n",
       "              ‘GRanges()’.\n",
       "\n",
       "          Because of this behavior, ‘GRanges(x)’ is equivalent to\n",
       "          ‘as(x, \"GRanges\")’.\n",
       "\n",
       "_\bC_\bo_\be_\br_\bc_\bi_\bo_\bn:\n",
       "\n",
       "     In the code snippets below, ‘x’ is a GRanges object.\n",
       "\n",
       "‘as(from, \"GRanges\")’: Creates a GRanges object from a character\n",
       "          vector, a factor, or a RangedData, RangesList, RleList, or\n",
       "          RleViewsList object.\n",
       "\n",
       "          When ‘from’ is a vector (or factor), each element in it must\n",
       "          represent a genomic range in format ‘chr1:2501-2800’\n",
       "          (unstranded range) or ‘chr1:2501-2800:+’ (stranded range).\n",
       "          ‘..’ is also supported as a separator between the start and\n",
       "          end positions. Strand can be ‘+’, ‘-’, ‘*’, or missing.  The\n",
       "          names on ‘from’ are propagated to the returned GRanges\n",
       "          object.  See ‘as.character()’ and ‘as.factor()’ below for the\n",
       "          reverse transformations.\n",
       "\n",
       "          Coercing a data.frame or DataFrame into a GRanges object is\n",
       "          also supported. See ‘makeGRangesFromDataFrame’ for the\n",
       "          details.\n",
       "\n",
       "‘as(from, \"RangedData\")’: Creates a RangedData object from a GRanges\n",
       "          object. The ‘strand’ and metadata columns become columns in\n",
       "          the result. The ‘seqlengths(from)’, ‘isCircular(from)’, and\n",
       "          ‘genome(from)’ vectors are stored in the metadata columns of\n",
       "          ‘ranges(rd)’.\n",
       "\n",
       "‘as(from, \"RangesList\")’: Creates a RangesList object from a GRanges\n",
       "          object. The ‘strand’ and metadata columns become _inner_\n",
       "          metadata columns (i.e. metadata columns on the ranges).  The\n",
       "          ‘seqlengths(from)’, ‘isCircular(from)’, and ‘genome(from)’\n",
       "          vectors become the metadata columns.\n",
       "\n",
       "‘as.character(x, ignore.strand=FALSE)’: Turn GRanges object ‘x’ into a\n",
       "          character vector where each range in ‘x’ is represented by a\n",
       "          string in format ‘chr1:2501-2800:+’. If ‘ignore.strand’ is\n",
       "          TRUE or if _all_ the ranges in ‘x’ are unstranded (i.e. their\n",
       "          strand is set to ‘*’), then all the strings in the output are\n",
       "          in format ‘chr1:2501-2800’.\n",
       "\n",
       "          The names on ‘x’ are propagated to the returned character\n",
       "          vector.  Its metadata (‘metadata(x)’) and metadata columns\n",
       "          (‘mcols(x)’) are ignored.\n",
       "\n",
       "          See ‘as(from, \"GRanges\")’ above for the reverse\n",
       "          transformation.\n",
       "\n",
       "‘as.factor(x)’: Equivalent to\n",
       "          \n",
       "            factor(as.character(x), levels=as.character(sort(unique(x))))\n",
       "\n",
       "          See ‘as(from, \"GRanges\")’ above for the reverse\n",
       "          transformation.\n",
       "\n",
       "          Note that ‘table(x)’ is supported on a GRanges object. It is\n",
       "          equivalent to, but much faster than, ‘table(as.factor(x))’.\n",
       "\n",
       "‘as.data.frame(x, row.names = NULL, optional = FALSE, ...)’: Creates a\n",
       "          data.frame with columns ‘seqnames’ (factor), ‘start’\n",
       "          (integer), ‘end’ (integer), ‘width’ (integer), ‘strand’\n",
       "          (factor), as well as the additional metadata columns stored\n",
       "          in ‘mcols(x)’. Pass an explicit ‘stringsAsFactors=TRUE/FALSE’\n",
       "          argument via ‘...’ to override the default conversions for\n",
       "          the metadata columns in ‘mcols(x)’.\n",
       "\n",
       "     In the code snippets below, ‘x’ is a Seqinfo object.\n",
       "\n",
       "‘as(x, \"GRanges\")’, ‘as(x, \"GenomicRanges\")’, ‘as(x, \"RangesList\")’:\n",
       "          Turns Seqinfo object ‘x’ (with no ‘NA’ lengths) into a\n",
       "          GRanges or RangesList.\n",
       "\n",
       "_\bA_\bc_\bc_\be_\bs_\bs_\bo_\br_\bs:\n",
       "\n",
       "     In the following code snippets, ‘x’ is a GRanges object.\n",
       "\n",
       "‘length(x)’: Get the number of elements.\n",
       "\n",
       "‘seqnames(x)’, ‘seqnames(x) <- value’: Get or set the sequence names.\n",
       "          ‘value’ can be an Rle object, a character vector, or a\n",
       "          factor.\n",
       "\n",
       "‘ranges(x)’, ‘ranges(x) <- value’: Get or set the ranges. ‘value’ can\n",
       "          be a Ranges object.\n",
       "\n",
       "‘names(x)’, ‘names(x) <- value’: Get or set the names of the elements.\n",
       "\n",
       "‘strand(x)’, ‘strand(x) <- value’: Get or set the strand. ‘value’ can\n",
       "          be an Rle object, character vector, or factor.\n",
       "\n",
       "‘mcols(x, use.names=FALSE)’, ‘mcols(x) <- value’: Get or set the\n",
       "          metadata columns.  If ‘use.names=TRUE’ and the metadata\n",
       "          columns are not ‘NULL’, then the names of ‘x’ are propagated\n",
       "          as the row names of the returned DataFrame object.  When\n",
       "          setting the metadata columns, the supplied value must be\n",
       "          ‘NULL’ or a data.frame-like object (i.e. DataTable or\n",
       "          data.frame) object holding element-wise metadata.\n",
       "\n",
       "‘elementMetadata(x)’, ‘elementMetadata(x) <- value’, ‘values(x)’,\n",
       "          ‘values(x) <- value’: Alternatives to ‘mcols’ functions.\n",
       "          Their use is discouraged.\n",
       "\n",
       "‘seqinfo(x)’, ‘seqinfo(x) <- value’: Get or set the information about\n",
       "          the underlying sequences.  ‘value’ must be a Seqinfo object.\n",
       "\n",
       "‘seqlevels(x)’, ‘seqlevels(x, force=FALSE) <- value’: Get or set the\n",
       "          sequence levels.  ‘seqlevels(x)’ is equivalent to\n",
       "          ‘seqlevels(seqinfo(x))’ or to ‘levels(seqnames(x))’, those 2\n",
       "          expressions being guaranteed to return identical character\n",
       "          vectors on a GRanges object.  ‘value’ must be a character\n",
       "          vector with no NAs.  See ‘?seqlevels’ for more information.\n",
       "\n",
       "‘seqlengths(x)’, ‘seqlengths(x) <- value’: Get or set the sequence\n",
       "          lengths.  ‘seqlengths(x)’ is equivalent to\n",
       "          ‘seqlengths(seqinfo(x))’.  ‘value’ can be a named\n",
       "          non-negative integer or numeric vector eventually with NAs.\n",
       "\n",
       "‘isCircular(x)’, ‘isCircular(x) <- value’: Get or set the circularity\n",
       "          flags.  ‘isCircular(x)’ is equivalent to\n",
       "          ‘isCircular(seqinfo(x))’.  ‘value’ must be a named logical\n",
       "          vector eventually with NAs.\n",
       "\n",
       "‘genome(x)’, ‘genome(x) <- value’: Get or set the genome identifier or\n",
       "          assembly name for each sequence.  ‘genome(x)’ is equivalent\n",
       "          to ‘genome(seqinfo(x))’.  ‘value’ must be a named character\n",
       "          vector eventually with NAs.\n",
       "\n",
       "‘seqlevelsStyle(x)’, ‘seqlevelsStyle(x) <- value’: Get or set the\n",
       "          seqname style for ‘x’.  See the seqlevelsStyle generic getter\n",
       "          and setter in the ‘GenomeInfoDb’ package for more\n",
       "          information.\n",
       "\n",
       "‘score(x), score(x) <- value’: Get or set the “score” column from the\n",
       "          element metadata.\n",
       "\n",
       "‘granges(x, use.mcols=FALSE)’: Gets a ‘GRanges’ with only the range\n",
       "          information from ‘x’, unless ‘use.mcols’ is ‘TRUE’, in which\n",
       "          case the metadata columns are also returned. Those columns\n",
       "          will include any \"extra column slots\" if ‘x’ is a specialized\n",
       "          ‘GenomicRanges’ derivative.\n",
       "\n",
       "_\bR_\ba_\bn_\bg_\be_\bs _\bm_\be_\bt_\bh_\bo_\bd_\bs:\n",
       "\n",
       "     In the following code snippets, ‘x’ is a GRanges object.\n",
       "\n",
       "‘start(x)’, ‘start(x) <- value’: Get or set ‘start(ranges(x))’.\n",
       "\n",
       "‘end(x)’, ‘end(x) <- value’: Get or set ‘end(ranges(x))’.\n",
       "\n",
       "‘width(x)’, ‘width(x) <- value’: Get or set ‘width(ranges(x))’.\n",
       "\n",
       "_\bS_\bp_\bl_\bi_\bt_\bt_\bi_\bn_\bg _\ba_\bn_\bd _\bC_\bo_\bm_\bb_\bi_\bn_\bi_\bn_\bg:\n",
       "\n",
       "     In the code snippets below, ‘x’ is a GRanges object.\n",
       "\n",
       "‘append(x, values, after = length(x))’: Inserts the ‘values’ into ‘x’\n",
       "          at the position given by ‘after’, where ‘x’ and ‘values’ are\n",
       "          of the same class.\n",
       "\n",
       "‘c(x, ...)’: Combines ‘x’ and the GRanges objects in ‘...’ together.\n",
       "          Any object in ‘...’ must belong to the same class as ‘x’, or\n",
       "          to one of its subclasses, or must be ‘NULL’.  The result is\n",
       "          an object of the same class as ‘x’.\n",
       "\n",
       "‘c(x, ..., ignore.mcols=FALSE)’ If the ‘GRanges’ objects have metadata\n",
       "          columns (represented as one DataFrame per object), each such\n",
       "          DataFrame must have the same columns in order to combine\n",
       "          successfully. In order to circumvent this restraint, you can\n",
       "          pass in an ‘ignore.mcols=TRUE’ argument which will combine\n",
       "          all the objects into one and drop all of their metadata\n",
       "          columns.\n",
       "\n",
       "‘split(x, f, drop=FALSE)’: Splits ‘x’ according to ‘f’ to create a\n",
       "          GRangesList object.  If ‘f’ is a list-like object then ‘drop’\n",
       "          is ignored and ‘f’ is treated as if it was\n",
       "          ‘rep(seq_len(length(f)), sapply(f, length))’, so the returned\n",
       "          object has the same shape as ‘f’ (it also receives the names\n",
       "          of ‘f’).  Otherwise, if ‘f’ is not a list-like object, empty\n",
       "          list elements are removed from the returned object if ‘drop’\n",
       "          is ‘TRUE’.\n",
       "\n",
       "_\bS_\bu_\bb_\bs_\be_\bt_\bt_\bi_\bn_\bg:\n",
       "\n",
       "     In the code snippets below, ‘x’ is a GRanges object.\n",
       "\n",
       "‘x[i, j]’, ‘x[i, j] <- value’: Get or set elements ‘i’ with optional\n",
       "          metadata columns ‘mcols(x)[,j]’, where ‘i’ can be missing; an\n",
       "          NA-free logical, numeric, or character vector; or a 'logical'\n",
       "          Rle object.\n",
       "\n",
       "‘x[i,j] <- value’: Replaces elements ‘i’ and optional metadata columns\n",
       "          ‘j’ with ‘value’.\n",
       "\n",
       "‘head(x, n = 6L)’: If ‘n’ is non-negative, returns the first n elements\n",
       "          of the GRanges object.  If ‘n’ is negative, returns all but\n",
       "          the last ‘abs(n)’ elements of the GRanges object.\n",
       "\n",
       "‘rep(x, times, length.out, each)’: Repeats the values in ‘x’ through\n",
       "          one of the following conventions:\n",
       "\n",
       "          ‘times’ Vector giving the number of times to repeat each\n",
       "              element if of length ‘length(x)’, or to repeat the whole\n",
       "              vector if of length 1.\n",
       "\n",
       "          ‘length.out’ Non-negative integer. The desired length of the\n",
       "              output vector.\n",
       "\n",
       "          ‘each’ Non-negative integer.  Each element of ‘x’ is repeated\n",
       "              ‘each’ times.\n",
       "\n",
       "‘subset(x, subset)’: Returns a new object of the same class as ‘x’ made\n",
       "          of the subset using logical vector ‘subset’, where missing\n",
       "          values are taken as ‘FALSE’.\n",
       "\n",
       "‘tail(x, n = 6L)’: If ‘n’ is non-negative, returns the last n elements\n",
       "          of the GRanges object.  If ‘n’ is negative, returns all but\n",
       "          the first ‘abs(n)’ elements of the GRanges object.\n",
       "\n",
       "‘window(x, start = NA, end = NA, width = NA, frequency = NULL, delta =\n",
       "          NULL, ...)’: Extracts the subsequence window from the GRanges\n",
       "          object using:\n",
       "\n",
       "          ‘start’, ‘end’, ‘width’ The start, end, or width of the\n",
       "              window. Two of the three are required.\n",
       "\n",
       "          ‘frequency’, ‘delta’ Optional arguments that specify the\n",
       "              sampling frequency and increment within the window.\n",
       "\n",
       "          In general, this is more efficient than using ‘\"[\"’ operator.\n",
       "\n",
       "‘window(x, start = NA, end = NA, width = NA, keepLength = TRUE) <-\n",
       "          value’: Replaces the subsequence window specified on the left\n",
       "          (i.e. the subsequence in ‘x’ specified by ‘start’, ‘end’ and\n",
       "          ‘width’) by ‘value’.  ‘value’ must either be of class\n",
       "          ‘class(x)’, belong to a subclass of ‘class(x)’, be coercible\n",
       "          to ‘class(x)’, or be ‘NULL’.  If ‘keepLength’ is ‘TRUE’, the\n",
       "          elements of ‘value’ are repeated to create a GRanges object\n",
       "          with the same number of elements as the width of the\n",
       "          subsequence window it is replacing.  If ‘keepLength’ is\n",
       "          ‘FALSE’, this replacement method can modify the length of\n",
       "          ‘x’, depending on how the length of the left subsequence\n",
       "          window compares to the length of ‘value’.\n",
       "\n",
       "‘x$name’, ‘x$name <- value’: Shortcuts for ‘mcols(x)$name’ and\n",
       "          ‘mcols(x)$name <- value’, respectively. Provided as a\n",
       "          convenience, for GRanges objects *only*, and as the result of\n",
       "          strong popular demand.  Note that those methods are not\n",
       "          consistent with the other ‘$’ and ‘$<-’ methods in the\n",
       "          IRanges/GenomicRanges infrastructure, and might confuse some\n",
       "          users by making them believe that a GRanges object can be\n",
       "          manipulated as a data.frame-like object.  Therefore we\n",
       "          recommend using them only interactively, and we discourage\n",
       "          their use in scripts or packages. For the latter, use\n",
       "          ‘mcols(x)$name’ and ‘mcols(x)$name <- value’, instead of\n",
       "          ‘x$name’ and ‘x$name <- value’, respectively.\n",
       "\n",
       "     Note that a GRanges object can be used to as a subscript to subset\n",
       "     a list-like object that has names on it. In that case, the names\n",
       "     on the list-like object are interpreted as sequence names.  In the\n",
       "     code snippets below, ‘x’ is a list or List object with names on\n",
       "     it, and the subscript ‘gr’ is a GRanges object with all its\n",
       "     seqnames being valid ‘x’ names.\n",
       "\n",
       "‘x[gr]’: Return an object of the same class as ‘x’ and _parallel_ to\n",
       "          ‘gr’. More precisely, it's conceptually doing:\n",
       "          \n",
       "            lapply(gr, function(gr1) x[[seqnames(gr1)]][ranges(gr1)])\n",
       "\n",
       "_\bO_\bt_\bh_\be_\br _\bm_\be_\bt_\bh_\bo_\bd_\bs:\n",
       "\n",
       "‘show(x)’: By default the ‘show’ method displays 5 head and 5 tail\n",
       "          elements. This can be changed by setting the global options\n",
       "          ‘showHeadLines’ and ‘showTailLines’. If the object length is\n",
       "          less than (or equal to) the sum of these 2 options plus 1,\n",
       "          then the full object is displayed.  Note that these options\n",
       "          also affect the display of GAlignments and GAlignmentPairs\n",
       "          objects (defined in the ‘GenomicAlignments’ package), as well\n",
       "          as other objects defined in the ‘IRanges’ and ‘Biostrings’\n",
       "          packages (e.g.  IRanges and DNAStringSet objects).\n",
       "\n",
       "_\bA_\bu_\bt_\bh_\bo_\br(_\bs):\n",
       "\n",
       "     P. Aboyoun and H. Pages\n",
       "\n",
       "_\bS_\be_\be _\bA_\bl_\bs_\bo:\n",
       "\n",
       "        • ‘makeGRangesFromDataFrame’\n",
       "\n",
       "        • ‘seqinfo’\n",
       "\n",
       "        • GenomicRanges-comparison\n",
       "\n",
       "        • intra-range-methods\n",
       "\n",
       "        • inter-range-methods\n",
       "\n",
       "        • coverage-methods\n",
       "\n",
       "        • setops-methods\n",
       "\n",
       "        • findOverlaps-methods\n",
       "\n",
       "        • nearest-methods\n",
       "\n",
       "        • ‘absoluteRanges’\n",
       "\n",
       "        • ‘tileGenome’\n",
       "\n",
       "        • GRangesList-class\n",
       "\n",
       "        • Ranges-class\n",
       "\n",
       "        • Vector-class\n",
       "\n",
       "        • Rle-class\n",
       "\n",
       "        • DataFrame-class\n",
       "\n",
       "_\bE_\bx_\ba_\bm_\bp_\bl_\be_\bs:\n",
       "\n",
       "     ## ---------------------------------------------------------------------\n",
       "     ## CONSTRUCTION\n",
       "     ## ---------------------------------------------------------------------\n",
       "     ## Specifying the bare minimum i.e. seqnames and ranges only. The\n",
       "     ## GRanges object will have no names, no strand information, and no\n",
       "     ## metadata columns:\n",
       "     gr0 <- GRanges(Rle(c(\"chr2\", \"chr2\", \"chr1\", \"chr3\"), c(1, 3, 2, 4)),\n",
       "                    IRanges(1:10, width=10:1))\n",
       "     gr0\n",
       "     \n",
       "     ## Specifying names, strand, metadata columns. They can be set on an\n",
       "     ## existing object:\n",
       "     names(gr0) <- head(letters, 10)\n",
       "     strand(gr0) <- Rle(strand(c(\"-\", \"+\", \"*\", \"+\", \"-\")), c(1, 2, 2, 3, 2))\n",
       "     mcols(gr0)$score <- 1:10\n",
       "     mcols(gr0)$GC <- seq(1, 0, length=10)\n",
       "     gr0\n",
       "     \n",
       "     ## ... or specified at construction time:\n",
       "     gr <- GRanges(Rle(c(\"chr2\", \"chr2\", \"chr1\", \"chr3\"), c(1, 3, 2, 4)),\n",
       "                   IRanges(1:10, width=10:1, names=head(letters, 10)),\n",
       "                   Rle(strand(c(\"-\", \"+\", \"*\", \"+\", \"-\")), c(1, 2, 2, 3, 2)),\n",
       "                   score=1:10, GC=seq(1, 0, length=10))\n",
       "     stopifnot(identical(gr0, gr))\n",
       "     \n",
       "     ## Specifying the seqinfo. It can be set on an existing object:\n",
       "     seqinfo <- Seqinfo(paste0(\"chr\", 1:3), c(1000, 2000, 1500), NA, \"mock1\")\n",
       "     seqinfo(gr0) <- merge(seqinfo(gr0), seqinfo)\n",
       "     seqlevels(gr0) <- seqlevels(seqinfo)\n",
       "     \n",
       "     ## ... or specified at construction time:\n",
       "     gr <- GRanges(Rle(c(\"chr2\", \"chr2\", \"chr1\", \"chr3\"), c(1, 3, 2, 4)),\n",
       "                   IRanges(1:10, width=10:1, names=head(letters, 10)),\n",
       "                   Rle(strand(c(\"-\", \"+\", \"*\", \"+\", \"-\")), c(1, 2, 2, 3, 2)),\n",
       "                   score=1:10, GC=seq(1, 0, length=10),\n",
       "                   seqinfo=seqinfo)\n",
       "     stopifnot(identical(gr0, gr))\n",
       "     \n",
       "     ## ---------------------------------------------------------------------\n",
       "     ## COERCION\n",
       "     ## ---------------------------------------------------------------------\n",
       "     ## From GRanges:\n",
       "     as.character(gr)\n",
       "     as.factor(gr)\n",
       "     as.data.frame(gr)\n",
       "     \n",
       "     ## From character to GRanges:\n",
       "     x1 <- \"chr2:56-125\"\n",
       "     as(x1, \"GRanges\")\n",
       "     as(rep(x1, 4), \"GRanges\")\n",
       "     x2 <- c(A=x1, B=\"chr1:25-30:-\")\n",
       "     as(x2, \"GRanges\")\n",
       "     \n",
       "     ## From data.frame to GRanges:\n",
       "     df <- data.frame(chrom=\"chr2\", start=11:15, end=20:24)\n",
       "     gr3 <- as(df, \"GRanges\")\n",
       "     \n",
       "     ## Alternatively, coercion to GRanges can be done by just calling the\n",
       "     ## GRanges() constructor on the object to coerce:\n",
       "     gr1 <- GRanges(x1)  # same as as(x1, \"GRanges\")\n",
       "     gr2 <- GRanges(x2)  # same as as(x2, \"GRanges\")\n",
       "     gr3 <- GRanges(df)  # same as as(df, \"GRanges\")\n",
       "     \n",
       "     ## Sanity checks:\n",
       "     stopifnot(identical(as(x1, \"GRanges\"), gr1))\n",
       "     stopifnot(identical(as(x2, \"GRanges\"), gr2))\n",
       "     stopifnot(identical(as(df, \"GRanges\"), gr3))\n",
       "     \n",
       "     ## ---------------------------------------------------------------------\n",
       "     ## SUMMARIZING ELEMENTS\n",
       "     ## ---------------------------------------------------------------------\n",
       "     table(seqnames(gr))\n",
       "     table(strand(gr))\n",
       "     sum(width(gr))\n",
       "     table(gr)\n",
       "     summary(mcols(gr)[,\"score\"])\n",
       "     \n",
       "     ## The number of lines displayed in the 'show' method are controlled\n",
       "     ## with two global options:\n",
       "     longGR  <- sample(gr, 25, replace=TRUE)\n",
       "     longGR\n",
       "     options(showHeadLines=7)\n",
       "     options(showTailLines=2)\n",
       "     longGR\n",
       "     \n",
       "     ## Revert to default values\n",
       "     options(showHeadLines=NULL)\n",
       "     options(showTailLines=NULL)\n",
       "     \n",
       "     ## ---------------------------------------------------------------------\n",
       "     ## RENAMING THE UNDERLYING SEQUENCES\n",
       "     ## ---------------------------------------------------------------------\n",
       "     seqlevels(gr)\n",
       "     seqlevels(gr) <- sub(\"chr\", \"Chrom\", seqlevels(gr))\n",
       "     gr\n",
       "     seqlevels(gr) <- sub(\"Chrom\", \"chr\", seqlevels(gr)) # revert\n",
       "     \n",
       "     ## ---------------------------------------------------------------------\n",
       "     ## COMBINING OBJECTS\n",
       "     ## ---------------------------------------------------------------------\n",
       "     gr2 <- GRanges(seqnames=Rle(c('chr1', 'chr2', 'chr3'), c(3, 3, 4)),\n",
       "                    IRanges(1:10, width=5),\n",
       "                    strand='-',\n",
       "                    score=101:110, GC=runif(10),\n",
       "                    seqinfo=seqinfo)\n",
       "     gr3 <- GRanges(seqnames=Rle(c('chr1', 'chr2', 'chr3'), c(3, 4, 3)),\n",
       "                    IRanges(101:110, width=10),\n",
       "                    strand='-',\n",
       "                    score=21:30,\n",
       "                    seqinfo=seqinfo)\n",
       "     some.gr <- c(gr, gr2)\n",
       "      \n",
       "     ## all.gr <- c(gr, gr2, gr3) ## (This would fail)\n",
       "     all.gr <- c(gr, gr2, gr3, ignore.mcols=TRUE)\n",
       "     \n",
       "     ## ---------------------------------------------------------------------\n",
       "     ## USING A GRANGES OBJECT AS A SUBSCRIPT TO SUBSET ANOTHER OBJECT\n",
       "     ## ---------------------------------------------------------------------\n",
       "     ## Subsetting *by* a GRanges subscript is supported only if the object\n",
       "     ## to subset is a named list-like object:\n",
       "     x <- RleList(chr1=101:120, chr2=2:-8, chr3=31:40)\n",
       "     x[gr]\n",
       "     "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# how to use GenomicRanges\n",
    "# 범위대로 잘라서 그때의 모습을 볼 수 있도록 : GenomicRanges\n",
    "?GenomicRanges"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## DNA /amino acid sequence from FASTA files\n",
    "***Biostrings*** classes (e.g., DNAStringSet) are used to represent DNA or amino acid sequences. In the example below we will construct a DNAString and show some manipulations."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "13"
      ],
      "text/latex": [
       "13"
      ],
      "text/markdown": [
       "13"
      ],
      "text/plain": [
       "[1] 13"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "  13-letter \"DNAString\" instance\n",
       "seq: TTGAAAA-CTC-N"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "library(Biostrings)\n",
    "d <- DNAString(\"TTGAAAA-CTC-N\")\n",
    "length(d)  #no of letters in the DNAString\n",
    "d"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We will download all Homo sapiens cDNA sequences from the FASTA file ‘Homo_sapiens.GRCh38.cdna.all.fa’ from Ensembl using ***AnnotationHub***."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      ": 'AnnotationHub' database may not be current\n",
      "  database: ‘/home/madbunny/.AnnotationHub/annotationhub.sqlite3’\n",
      "  reason: Problem with the SSL CA cert (path? access rights?)"
     ]
    }
   ],
   "source": [
    "#biocLite(\"devtools\")\n",
    "#biocLite(\"hadley/httr\")\n",
    "library(AnnotationHub)\n",
    "#removeCache()\n",
    "# curl::curl_fetch_disk(\"www.google.com\", tempfile())\n",
    "ah <- AnnotationHub()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnotationHub with 35307 records\n",
       "# snapshotDate(): 2016-03-07 \n",
       "# $dataprovider: BroadInstitute, UCSC, Ensembl, NCBI, Haemcode, Inparanoid8,...\n",
       "# $species: Homo sapiens, Mus musculus, Bos taurus, Pan troglodytes, Danio r...\n",
       "# $rdataclass: GRanges, FaFile, BigWigFile, OrgDb, ChainFile, Inparanoid8Db,...\n",
       "# additional mcols(): taxonomyid, genome, description, tags, sourceurl,\n",
       "#   sourcetype \n",
       "# retrieve records with, e.g., 'object[[\"AH2\"]]' \n",
       "\n",
       "            title                                               \n",
       "  AH2     | Ailuropoda_melanoleuca.ailMel1.69.dna.toplevel.fa   \n",
       "  AH3     | Ailuropoda_melanoleuca.ailMel1.69.dna_rm.toplevel.fa\n",
       "  AH4     | Ailuropoda_melanoleuca.ailMel1.69.dna_sm.toplevel.fa\n",
       "  AH5     | Ailuropoda_melanoleuca.ailMel1.69.ncrna.fa          \n",
       "  AH6     | Ailuropoda_melanoleuca.ailMel1.69.pep.all.fa        \n",
       "  ...       ...                                                 \n",
       "  AH49436 | Xiphophorus_maculatus.Xipmac4.4.2.dna_rm.toplevel.fa\n",
       "  AH49437 | Xiphophorus_maculatus.Xipmac4.4.2.dna_sm.toplevel.fa\n",
       "  AH49438 | Xiphophorus_maculatus.Xipmac4.4.2.dna.toplevel.fa   \n",
       "  AH49439 | Xiphophorus_maculatus.Xipmac4.4.2.ncrna.fa          \n",
       "  AH49440 | Xiphophorus_maculatus.Xipmac4.4.2.pep.all.fa        "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ah"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "class: FaFile \n",
       "path: /home/madbunny/.AnnotationHub/22617\n",
       "index: /home/madbunny/.AnnotationHub/25666\n",
       "isOpen: FALSE \n",
       "yieldSize: NA "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ah2 <- query(ah, c(\"fasta\", \"homo sapiens\", \"Ensembl\"))\n",
    "fa <- ah2[[\"AH18522\"]]\n",
    "fa"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We will open the file and get the sequences and widths of the records in the the fasta file using ***Rsamtools***."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 170893 ranges and 0 metadata columns:\n",
       "                  seqnames    ranges strand\n",
       "                     <Rle> <IRanges>  <Rle>\n",
       "       [1] ENST00000434970   [1,  9]      *\n",
       "       [2] ENST00000415118   [1,  8]      *\n",
       "       [3] ENST00000448914   [1, 13]      *\n",
       "       [4] ENST00000431870   [1, 16]      *\n",
       "       [5] ENST00000414852   [1, 16]      *\n",
       "       ...             ...       ...    ...\n",
       "  [170889] ENST00000444082 [1, 3808]      *\n",
       "  [170890] ENST00000615390 [1,  859]      *\n",
       "  [170891] ENST00000512197 [1,  115]      *\n",
       "  [170892] ENST00000414573 [1,  204]      *\n",
       "  [170893] ENST00000428912 [1,  797]      *\n",
       "  -------\n",
       "  seqinfo: 170893 sequences from an unspecified genome"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "library(Rsamtools)\n",
    "idx <- scanFaIndex(fa)\n",
    "idx"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The information is returned as a GRanges object. ***getSeq()*** returns the sequences indicated by param as a DNAStringSet instance."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "  A DNAStringSet instance of length 7\n",
       "     width seq                                              names               \n",
       "[1] 101518 GCAGTCGTGCATTCCCAGCCTCG...GCAAGCTATCTGCACCTCAAAA ENST00000342992\n",
       "[2]  82029 GAGCAGTCGTGCATTCCCAGCCT...GCAAGCTATCTGCACCTCAAAA ENST00000460472\n",
       "[3] 109224 GAGCAGTCGTGCATTCCCAGCCT...GCAAGCTATCTGCACCTCAAAA ENST00000589042\n",
       "[4] 104301 GAGCAGTCGTGCATTCCCAGCCT...GCAAGCTATCTGCACCTCAAAA ENST00000591111\n",
       "[5] 104301 GAGCAGTCGTGCATTCCCAGCCT...GCAAGCTATCTGCACCTCAAAA ENST00000615779\n",
       "[6]  82605 GAGCAGTCGTGCATTCCCAGCCT...GCAAGCTATCTGCACCTCAAAA ENST00000342175\n",
       "[7]  82404 GAGCAGTCGTGCATTCCCAGCCT...GCAAGCTATCTGCACCTCAAAA ENST00000359218"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "long <- idx[width(idx) > 82000]\n",
    "getSeq(fa, param=long)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***BSgenome*** packages inside Bioconductor contain whole genome sequences as distributed by ENSEMBL, NCBI and others. In this next example we will load the whole genome sequence for Homo sapiens from UCSC’s hg19 build, and calculate the GC content across chromosome 14."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<thead><tr><th scope=col>G|C</th></tr></thead>\n",
       "<tbody>\n",
       "\t<tr><td>0.336276</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "\\begin{tabular}{l}\n",
       " G|C\\\\\n",
       "\\hline\n",
       "\t 0.336276\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "0.336276047386882"
      ],
      "text/plain": [
       "          G|C\n",
       "[1,] 0.336276"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "library(BSgenome.Hsapiens.UCSC.hg19)\n",
    "\n",
    "chr14_range = GRanges(\"chr14\", IRanges(1, seqlengths(Hsapiens)[\"chr14\"]))\n",
    "chr14_dna <- getSeq(Hsapiens, chr14_range)\n",
    "letterFrequency(chr14_dna, \"GC\", as.prob=TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.2.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
